---
title: "CCA somatic enhancers, group-specific"
output: 
  html_notebook:
    toc: true
    toc_float: true

---



  
# CCA somatic group-specific enhancers, validate by expr & cell-lines

  
  

```{r include=F}

# options(scipen=0)
rm(list=ls())
library(DiffBind, warn.conflicts=F, quietly=T)
library(beeswarm)
library(msigdbr)
library(fgsea)
library(gplots)
library(clusterProfiler)
library(biomaRt)
library(ggplot2)
source('analysis_cca_enhancers_functions.R')

# # read diffbind results
# load(file="Rdata_cca_k27ac_3c_cells_diffbind.Rdata")
# rm(K27ac_cells_dba, K27ac_cells_enhsup_tisspeaks_dba)
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts1.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts2.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts3.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts4.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts5.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts1_diffexpr.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts2_diffexpr.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts3_diffexpr.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts4_diffexpr.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts5_diffexpr.Rdata")

# read mapping
target_map = read.table("target_gene_mapping/20190628_enhancer_target_map_noTAD_CPLW_rho.3_liberal.txt", header=T, stringsAsFactors = F) 


# read differential enhancers from tissue
# load("C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/Rdata_cca_k27ac_2b_diffbind_diffenh.Rdata")

# read expression data: cellline counts. Ensure samples have same IDs as chipseq data!
expr_counts_cell_data = read.table("expression/counts_matrix_celllines/counts_matrix-genes.txt", stringsAsFactors = F, header=T, check.names = F, sep="\t") # 26589 rows, all unique gene names.
rownames(expr_counts_cell_data) = expr_counts_cell_data$gene_name
expr_counts_cell_data = expr_counts_cell_data[, !colnames(expr_counts_cell_data) %in% c("gene_id", "gene_name")]
expr_counts_cell_data = round(expr_counts_cell_data)
expr_counts_cell_data = expr_counts_cell_data[, grep("H69-KO|S5-EV|S5-MUT|S5-WT", colnames(expr_counts_cell_data), invert = T)]
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="Egi-1-1"] = "EGI1_R0"
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="Egi-1-2"] = "EGI1_R1"
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="H69-1"] = "H69_R0"
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="H69-2"] = "H69_R1"
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="HUCCT"] = "HUCCT_R1"
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="KKU100"] = "KKU100_R1"
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="M213"] = "M214"



# read expression data: tissue counts. Ensure samples have same IDs as chipseq data!
expr_counts_tiss_data = read.table("expression/counts_matrix_tissue_CCA_HCC/counts_matrix-genes.txt", stringsAsFactors = F, header=T, check.names = F, sep="\t") # 26589 rows, all unique gene names.
rownames(expr_counts_tiss_data) = expr_counts_tiss_data$gene_name
expr_counts_tiss_data = expr_counts_tiss_data[, !colnames(expr_counts_tiss_data) %in% c("gene_id", "gene_name")]
expr_counts_tiss_data = round(expr_counts_tiss_data)


# read mSigDB genesets
msigdb_H = msigdbr(species = "Homo sapiens", category = "H") # 50 genesets
msigdb_R = msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME") # 674 genesets
msigdb_C = msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CGP") # 3433 genesets
msigdb_H_list = msigdb_H %>% split(x = .$gene_symbol, f = .$gs_name)
msigdb_HRC_list = rbind(rbind(msigdb_H, msigdb_R), msigdb_C) %>% split(x = .$gene_symbol, f = .$gs_name)
msigdb_H_term2gene = data.frame(TERM=msigdb_H$gs_name, GENE=msigdb_H$gene_symbol, stringsAsFactors = F)
msigdb_R_term2gene = data.frame(TERM=msigdb_R$gs_name, GENE=msigdb_R$gene_symbol, stringsAsFactors = F)
msigdb_C_term2gene = data.frame(TERM=msigdb_C$gs_name, GENE=msigdb_C$gene_symbol, stringsAsFactors = F)
msigdb_H_term2gene = rbind(msigdb_H_term2gene, data.frame(TERM="ALL_GENES", GENE=unique(na.omit(target_map$gene)), stringsAsFactors = F)) # clusterProfiler considers universe = genes in genesets
msigdb_R_term2gene = rbind(msigdb_R_term2gene, data.frame(TERM="ALL_GENES", GENE=unique(na.omit(target_map$gene)), stringsAsFactors = F))
msigdb_C_term2gene = rbind(msigdb_C_term2gene, data.frame(TERM="ALL_GENES", GENE=unique(na.omit(target_map$gene)), stringsAsFactors = F))

# custom genesets
msigdb_HRC_list$YONGCHERNHAN_STING = unique(c(msigdb_HRC_list$ISHIKAWA_STING_SIGNALING, "MB21D1", "TMEM173", "IRF3", "TBK1"))
msigdb_C_term2gene = rbind(msigdb_C_term2gene, data.frame(TERM="YONGCHERNHAN_STING", GENE=msigdb_HRC_list$YONGCHERNHAN_STING, stringsAsFactors = F))


# read biomart gene descriptions
biomart_ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")


get_genesets = function(gene, pathways=c("H","R")) {
  ann = NULL
  if ("H" %in% pathways) {
    ann = c(ann, msigdb_H_term2gene[msigdb_H_term2gene$GENE==gene,"TERM"])
  }
  if ("R" %in% pathways) {
    ann = c(ann, msigdb_R_term2gene[msigdb_R_term2gene$GENE==gene,"TERM"])
  }
  ann = unique(ann)
  ann = ann[ann!="ALL_GENES"]
  ann = gsub("HALLMARK", "H", ann)
  ann = gsub("REACTOME", "R", ann)
  ann = sapply(ann, function(x){substr(x, 1, 30)})
  paste0(ann, collapse=",")
}
get_gene_descr = function(gene){
  descr = getBM(attributes=c('description'), filters='external_gene_name', values=gene, mart=biomart_ensembl)[1,1]
  if (is.na(descr)) { "" 
  } else {  strsplit(descr, split=" [", fixed=T)[[1]][1] }
}


# read mutations
muts_targetedgenes = read.table(file="genomic alterations/2019.03.07_mutations_targeted.txt", sep="\t", header=T, stringsAsFactors = F) # 5368 rows, 471 samples total (only 428 samples with muts): 71 wgs (all with muts), 388 prev (346 with muts), 12 taiwan (11 with muts)



```



```{r}
### For plotting 



### read enhancer data 
load("Rdata_cca_k27ac_0_prelim_final.Rdata")
rm(K27ac_enhsup_dba, superenhancers)
K27ac_dba = dba(DBA=K27ac_dba, mask=!K27ac_dba$masks$Normal_liver) # 71 samples,   sites

## enhancer signals using DBA_SCORE_TMM_MINUS_FULL. This accounts for lib size, but not enhancer size.
# Note that this score is already calculated in K27ac_dba, but if recalculate it after removing liver, then it'll change!
# K27ac_dba = dba.count(K27ac_dba, peaks=NULL, score=DBA_SCORE_TMM_MINUS_FULL)
K27ac_enhancer_signalsTMM = dba.peakset(K27ac_dba, bRetrieve = T, DataType = DBA_DATA_FRAME) # 474060 enhancers
rownames(K27ac_enhancer_signalsTMM) = paste0(K27ac_enhancer_signalsTMM$CHR, ":", K27ac_enhancer_signalsTMM$START, "-", K27ac_enhancer_signalsTMM$END)
K27ac_enhancer_signalsTMM = K27ac_enhancer_signalsTMM[, 4:ncol(K27ac_enhancer_signalsTMM)]
# casting to dataframe may change some of the sample names, so rename them. First check that sample ordering is as expected
as.character(dba.show(K27ac_dba)$ID)
colnames(K27ac_enhancer_signalsTMM)
# rename samples
colnames(K27ac_enhancer_signalsTMM) = as.character(dba.show(K27ac_dba)$ID)

# split into groups
samples_data = read.table("2021.08.22 samples_info.txt", stringsAsFactors = F, header=T, sep="\t", check.names = F)
K27ac_enhancer_signalsTMM_Fluke = K27ac_enhancer_signalsTMM[, samples_data[samples_data$EGroup=="Fluke", "Sample"]]
K27ac_enhancer_signalsTMM_C4 = K27ac_enhancer_signalsTMM[, samples_data[samples_data$EGroup=="C4", "Sample"]]
K27ac_enhancer_signalsTMM_Mixed = K27ac_enhancer_signalsTMM[, samples_data[samples_data$EGroup=="Mixed", "Sample"]]
K27ac_enhancer_signalsTMM_Normal = K27ac_enhancer_signalsTMM[, samples_data[samples_data$EGroup=="Normal", "Sample"]]


# read expressions TPM data
expr_TPM_data = read.table("expression/counts_matrix_tissue_CCA_HCC/tpm_matrix-genes.txt", stringsAsFactors = F, header=T, check.names = F, sep="\t") # 24622 rows, all unique gene names.
rownames(expr_TPM_data) = expr_TPM_data$gene_name
expr_TPM_data = expr_TPM_data[, !colnames(expr_TPM_data) %in% c("gene_id", "gene_name")]
expr_TPM_data = expr_TPM_data[, colnames(expr_TPM_data) %in% as.character(dba.show(K27ac_dba)$ID)] # 24622 genes x 66 samples
# split into groups
expr_TPM_data_Fluke = expr_TPM_data[, intersect(colnames(expr_TPM_data), samples_data[samples_data$EGroup=="Fluke", "Sample"])]
expr_TPM_data_C4 = expr_TPM_data[, intersect(colnames(expr_TPM_data), samples_data[samples_data$EGroup=="C4", "Sample"])]
expr_TPM_data_Mixed = expr_TPM_data[, intersect(colnames(expr_TPM_data), samples_data[samples_data$EGroup=="Mixed", "Sample"])]
expr_TPM_data_Normal = expr_TPM_data[, intersect(colnames(expr_TPM_data), samples_data[samples_data$EGroup=="Normal", "Sample"])]

```




```{r}


somatic_enhancers_groupspec = list()

# just load
# load("Rdata_cca_k27ac_3b_somatic_enhancers_groupspec.Rdata")


```


# EGroup Fluke somatic enhancers
```{r}

load(file="Rdata_cca_k27ac_3_cells_diffbind.Rdata")
rm(K27ac_cells_dba, K27ac_cells_enhsup_tisspeaks_dba)
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts1.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts1_diffexpr.Rdata")
```


### EGroup Fluke vs EGroup C4 (w/ Norm ref), just validate with Normal
```{r fig.width=8}
contrast = 1

somatic_enhancers_groupspec[[contrast]] = validate_diff_enhancers_multicomp_exp_cellline(K27ac_dba.x.1=K27ac_dba_deseq_contrasts1, K27ac_dba.x.2=K27ac_dba_deseq_contrasts1, contrasts.x=c(1, 3),
                                              group_interest.x=c(1,1), method.x=DBA_DESEQ2, diffexprlist.x.1=diffexpr_list_contrasts1, diffexprlist.x.2=diffexpr_list_contrasts1,  
                                               K27ac_dba.y.1=K27ac_cells_tisspeaks_dba, K27ac_dba.y.2=K27ac_cells_tisspeaks_dba, contrasts.y=c(2,2), group_interest.y=c(1,1),
                                               method.y=DBA_DESEQ2, diffexprlist.y.1=diffexpr_cells_list, diffexprlist.y.2=diffexpr_cells_list,
                                               corr_data=target_map, enh_present_thr=.5, filterEnh=NULL, pathways_list=msigdb_H_list, enh_p_or_q="q", exp_p_or_q="p")
names(somatic_enhancers_groupspec)[[contrast]] = "Tumor_EGroupFluke_vs_Tumor_EGroupC4"


clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$gain_exp_enhgain$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$lost_exp_enhlost$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}


### gsea
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts1, K27ac_dba2=K27ac_dba_deseq_contrasts1, contrasts=c(1,3), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts1, diffexprlist2=diffexpr_list_contrasts1, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=2, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }


```








### EGroup Fluke vs EGroup Mixed (w/ Norm ref), just validate vs normal
```{r fig.width=8}
contrast = 2

somatic_enhancers_groupspec[[contrast]] = validate_diff_enhancers_multicomp_exp_cellline(K27ac_dba.x.1=K27ac_dba_deseq_contrasts1, K27ac_dba.x.2=K27ac_dba_deseq_contrasts1, contrasts.x=c(1, 4),
                                              group_interest.x=c(1,1), method.x=DBA_DESEQ2, diffexprlist.x.1=diffexpr_list_contrasts1, diffexprlist.x.2=diffexpr_list_contrasts1,  
                                               K27ac_dba.y.1=K27ac_cells_tisspeaks_dba, K27ac_dba.y.2=K27ac_cells_tisspeaks_dba, contrasts.y=c(2,2), group_interest.y=c(1,1),
                                               method.y=DBA_DESEQ2, diffexprlist.y.1=diffexpr_cells_list, diffexprlist.y.2=diffexpr_cells_list,
                                               corr_data=target_map, enh_present_thr=.5, filterEnh=NULL, pathways_list=msigdb_H_list, enh_p_or_q="q", exp_p_or_q="p")

names(somatic_enhancers_groupspec)[[contrast]] = "Tumor_EGroupFluke_vs_Tumor_EGroupMixed"


clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$gain_exp_enhgain$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}


clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$lost_exp_enhlost$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}



### gsea
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts1, K27ac_dba2=K27ac_dba_deseq_contrasts1, contrasts=c(1,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts1, diffexprlist2=diffexpr_list_contrasts1, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=2, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }


```








### use EGroup Fluke vs C4 & Mixed
```{r fig.width=8}

# gain
EGroupFluke_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain$gene, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain$gene)
analyze_geneset_enrichment(genes_interest=unique(EGroupFluke_intersect), genes_univ=unique(target_map$gene), pathways_list=msigdb_H_list)

clusprof_enrich_H = enricher(gene=unique(EGroupFluke_intersect), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { 
  dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}


# lost
EGroupFluke_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$lost_exp_enhlost$gene, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$lost_exp_enhlost$gene)
analyze_geneset_enrichment(genes_interest=unique(EGroupFluke_intersect), genes_univ=unique(target_map$gene), pathways_list=msigdb_H_list)

clusprof_enrich_H = enricher(gene=unique(EGroupFluke_intersect), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { 
  dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_bw(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}


# gsea
### Filter by expression & cell line existence & enhancer change
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts1, K27ac_dba2=K27ac_dba_deseq_contrasts1, contrasts=c(1,3,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts1, diffexprlist2=diffexpr_list_contrasts1, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=2, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { 
  dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_bw(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { 
  dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_bw(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
  }
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }



```






```{r}
### validation statistics of gained enhancers
EGroupFluke_intersect_enh1 = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain$enhancer)

EGroupFluke_intersect_enh2 = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp$enhancer)

EGroupFluke_intersect_enh3 = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enh$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enh$enhancer)

EGroupFluke_intersect_enh4 = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain$enhancer)

print(paste0("Gain: ", length(EGroupFluke_intersect_enh1)))
print(paste0("Gain & exp: ", length(EGroupFluke_intersect_enh2)))
print(paste0("Gain & exp & enh: ", length(EGroupFluke_intersect_enh3)))
print(paste0("Gain & exp & enh gain: ", length(EGroupFluke_intersect_enh4)))

```

```{r}

#### check how many validated enhancers are estro genes
estro_genes = union(msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_EARLY, msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_LATE)

# get enhancer-gene pairs
somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain$enhancergene = paste0(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain$enhancer, '|', somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain$gene)

somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain$enhancergene = paste0(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain$enhancer, '|', somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain$gene)

EGroupFluke_intersect_enhgene = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain$enhancergene, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain$enhancergene)



#### check how many validated enhancers are estro genes
gained_estro = data.frame(enhancer=unlist(lapply(strsplit(EGroupFluke_intersect_enhgene, split='|', fixed=T), function(x){x[1]})),
                gene=unlist(lapply(strsplit(EGroupFluke_intersect_enhgene, split='|', fixed=T), function(x){x[2]})), 
                enhancergene=EGroupFluke_intersect_enhgene, stringsAsFactors = F)
gained_estro$estro = unlist(lapply(gained_estro$gene, function(x){x %in% estro_genes}))
gained_estro = gained_estro[gained_estro$estro,]
# add enhancer target correlation
target_map$enhancergene = paste0(target_map$enhancer, '|', target_map$gene)
gained_estro = merge(gained_estro, target_map[, c("enhancergene", "rho", "promK27ac_rho")], by="enhancergene")

print("Validated enhancer-genes assoc with estro")
print(gained_estro)
print(paste0("Num validated enhancers assoc with estro genes: ", length(unique(gained_estro$enhancer))))

### check how many validated enhancers' genes are estro genes
print (paste0("Num validated enhancers' genes are estro genes: ", length(unique(gained_estro$gene))))

```

```{r fig.width=4, fig.asp=1.5}
### boxplot enhancer & expression for selected genes: SFN, SLC2A1, SORD, MLPH, MYB, ELF3, MAPK13, TMPRSS3, KRT19, CLIC3, KLF5
gained_Fluke = data.frame(enhancer=unlist(lapply(strsplit(EGroupFluke_intersect_enhgene, split='|', fixed=T), function(x){x[1]})),
                gene=unlist(lapply(strsplit(EGroupFluke_intersect_enhgene, split='|', fixed=T), function(x){x[2]})), 
                enhancergene=EGroupFluke_intersect_enhgene, stringsAsFactors = F)
target_map$enhancergene = paste0(target_map$enhancer, '|', target_map$gene)
gained_Fluke = merge(gained_Fluke, target_map[, c("enhancergene", "rho", "promK27ac_rho")], by="enhancergene")

for (gg in c("SFN", "SLC2A1", "SORD", "MLPH", "MYB", "ELF3", "MAPK13", "TMPRSS3", "KRT19", "CLIC3", "KLF5")) {
  print(gg)
  zz = gained_Fluke[gained_Fluke$gene==gg, ]
  ee = zz[order(zz$rho, decreasing = T),][1, "enhancer"]
  boxplot(as.numeric(K27ac_enhancer_signalsTMM_Normal[ee,]),
          as.numeric(K27ac_enhancer_signalsTMM_Fluke[ee,]),
          as.numeric(K27ac_enhancer_signalsTMM_C4[ee,]),
          as.numeric(K27ac_enhancer_signalsTMM_Mixed[ee,]),
          at=c(1,2,3,4), notch=F, col=c("grey", "cornflowerblue", "red2","green"), main=paste0(gg, " enhancer"))
  boxplot(as.numeric(expr_TPM_data_Normal[gg,]),
          as.numeric(expr_TPM_data_Fluke[gg,]),
          as.numeric(expr_TPM_data_C4[gg,]),
          as.numeric(expr_TPM_data_Mixed[gg,]),
          at=c(1,2,3,4), notch=F, col=c("grey", "cornflowerblue", "red2","green"), main=paste0(gg, " expression"))
}
```



```{r}



### get gained estrogen enhancers that are mtorc1 (from jinghan)
gained_estro_mtorc_JH = gained_estro[gained_estro$gene %in% c("SFN", "SLC2A1", "SORD", "MLPH", "MYB", "ELF3"),]
gained_estro_mtorc_JH_vsC4stats = somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain[somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain$enhancergene %in% gained_estro_mtorc_JH$enhancergene,]
gained_estro_mtorc_JH_vsC4stats = gained_estro_mtorc_JH_vsC4stats[!is.na(gained_estro_mtorc_JH_vsC4stats$gene),]
gained_estro_mtorc_JH_vsC4stats = gained_estro_mtorc_JH_vsC4stats[order(gained_estro_mtorc_JH_vsC4stats$gene, gained_estro_mtorc_JH_vsC4stats$enhancer),]

gained_estro_mtorc_JH_vsMixedstats = somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain[somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain$enhancergene %in% gained_estro_mtorc_JH$enhancergene,]
gained_estro_mtorc_JH_vsMixedstats = gained_estro_mtorc_JH_vsMixedstats[!is.na(gained_estro_mtorc_JH_vsMixedstats$gene),]
gained_estro_mtorc_JH_vsMixedstats = gained_estro_mtorc_JH_vsMixedstats[order(gained_estro_mtorc_JH_vsMixedstats$gene, gained_estro_mtorc_JH_vsMixedstats$enhancer),]

# add enhancer-gene correlation stats
target_map$enhancergene = paste0(target_map$enhancer, '|', target_map$gene)
gained_estro_mtorc_JH_vsC4stats = merge(gained_estro_mtorc_JH_vsC4stats, target_map[, c("enhancergene", "rho", "promK27ac_rho")], by="enhancergene")
gained_estro_mtorc_JH_vsMixedstats = merge(gained_estro_mtorc_JH_vsMixedstats, target_map[, c("enhancergene", "rho", "promK27ac_rho")], by="enhancergene")

print(gained_estro_mtorc_JH_vsC4stats)
print(gained_estro_mtorc_JH_vsMixedstats)

```

```{r}
# print gained enhancers in bed format to put in ucsc genome browser multiview
zz = gained_estro_mtorc_JH_vsC4stats[gained_estro_mtorc_JH_vsC4stats$gene=="SFN", ]
zz = zz[order(zz$rho, decreasing = T),][1:min(5, nrow(zz)), ]
bed_coords = data.frame(chr=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){x[1]})), 
                        start=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[2])-500})),
                        end=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[3])+500})), stringsAsFactors = F)


zz = gained_estro_mtorc_JH_vsC4stats[gained_estro_mtorc_JH_vsC4stats$gene=="SLC2A1", ]
zz = zz[order(zz$rho, decreasing = T),][1:min(5, nrow(zz)), ]
bed_coords = data.frame(chr=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){x[1]})), 
                        start=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[2])-500})),
                        end=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[3])+500})), stringsAsFactors = F)



zz = gained_estro_mtorc_JH_vsC4stats[gained_estro_mtorc_JH_vsC4stats$gene=="MYB", ]
zz = zz[order(zz$rho, decreasing = T),][1:min(5, nrow(zz)), ]
bed_coords = data.frame(chr=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){x[1]})), 
                        start=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[2])-500})),
                        end=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[3])+500})), stringsAsFactors = F)


zz = gained_estro_mtorc_JH_vsC4stats[gained_estro_mtorc_JH_vsC4stats$gene=="ELF3", ]
zz = zz[order(zz$rho, decreasing = T),][1:min(5, nrow(zz)), ]
bed_coords = data.frame(chr=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){x[1]})), 
                        start=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[2])-500})),
                        end=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[3])+500})), stringsAsFactors = F)


```






```{r}
### validation statistics of lost enhancers
EGroupFluke_intersect_enh1 = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$lost$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$lost$enhancer)

EGroupFluke_intersect_enh2 = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$lost_exp$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$lost_exp$enhancer)

EGroupFluke_intersect_enh3 = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$lost_exp_enh$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$lost_exp_enh$enhancer)

EGroupFluke_intersect_enh4 = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$lost_exp_enhlost$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$lost_exp_enhlost$enhancer)

print(paste0("lost: ", length(EGroupFluke_intersect_enh1)))
print(paste0("lost & exp: ", length(EGroupFluke_intersect_enh2)))
print(paste0("lost & exp & enh: ", length(EGroupFluke_intersect_enh3)))
print(paste0("lost & exp & enh lost: ", length(EGroupFluke_intersect_enh4)))

```




# EGroup C4 somatic enhancers
```{r}

load(file="Rdata_cca_k27ac_3_cells_diffbind.Rdata")
rm(K27ac_cells_dba, K27ac_cells_enhsup_tisspeaks_dba)
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts2.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts2_diffexpr.Rdata")
```


### EGroup C4 vs Fluke (w/ Norm ref) - just validate vs normal
```{r fig.width=8}
contrast = 3

somatic_enhancers_groupspec[[contrast]] = validate_diff_enhancers_multicomp_exp_cellline(K27ac_dba.x.1=K27ac_dba_deseq_contrasts2, K27ac_dba.x.2=K27ac_dba_deseq_contrasts2, contrasts.x=c(1, 3),
                                              group_interest.x=c(1,1), method.x=DBA_DESEQ2, diffexprlist.x.1=diffexpr_list_contrasts2, diffexprlist.x.2=diffexpr_list_contrasts2,  
                                               K27ac_dba.y.1=K27ac_cells_tisspeaks_dba, K27ac_dba.y.2=K27ac_cells_tisspeaks_dba, contrasts.y=c(3,3), group_interest.y=c(1,1),
                                               method.y=DBA_DESEQ2, diffexprlist.y.1=diffexpr_cells_list, diffexprlist.y.2=diffexpr_cells_list,
                                               corr_data=target_map, enh_present_thr=.5, filterEnh=NULL, pathways_list=msigdb_H_list, enh_p_or_q="q", exp_p_or_q="p")

names(somatic_enhancers_groupspec)[[contrast]] = "Tumor_EGroupC4_vs_Tumor_EGroupFluke"

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$gain_exp_enhgain$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}


clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$lost_exp_enhlost$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}





### gsea
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts2, K27ac_dba2=K27ac_dba_deseq_contrasts2, contrasts=c(1,3), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts2, diffexprlist2=diffexpr_list_contrasts2, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }


```





### EGroup C4 vs Mixed (w/ Norm ref) - just validate vs normal
```{r fig.width=8}
contrast = 4

somatic_enhancers_groupspec[[contrast]] = validate_diff_enhancers_multicomp_exp_cellline(K27ac_dba.x.1=K27ac_dba_deseq_contrasts2, K27ac_dba.x.2=K27ac_dba_deseq_contrasts2, contrasts.x=c(1, 4),
                                              group_interest.x=c(1,1), method.x=DBA_DESEQ2, diffexprlist.x.1=diffexpr_list_contrasts2, diffexprlist.x.2=diffexpr_list_contrasts2,  
                                               K27ac_dba.y.1=K27ac_cells_tisspeaks_dba, K27ac_dba.y.2=K27ac_cells_tisspeaks_dba, contrasts.y=c(3,3), group_interest.y=c(1,1),
                                               method.y=DBA_DESEQ2, diffexprlist.y.1=diffexpr_cells_list, diffexprlist.y.2=diffexpr_cells_list,
                                               corr_data=target_map, enh_present_thr=.5, filterEnh=NULL, pathways_list=msigdb_H_list, enh_p_or_q='q', exp_p_or_q="p")

names(somatic_enhancers_groupspec)[[contrast]] = "Tumor_EGroupC4_vs_Tumor_EGroupMixed"

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$gain_exp_enhgain$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$lost_exp_enhlost$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}




### gsea
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts2, K27ac_dba2=K27ac_dba_deseq_contrasts2, contrasts=c(1,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts2, diffexprlist2=diffexpr_list_contrasts2, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }

```





### EGroup C4 vs Fluke & Mixed
```{r fig.width=8}

# gain
EGroupC4_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain$gene)
analyze_geneset_enrichment(genes_interest=unique(EGroupC4_intersect), genes_univ=unique(target_map$gene), pathways_list=msigdb_H_list)

clusprof_enrich_H = enricher(gene=unique(EGroupC4_intersect), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

# lost
EGroupC4_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$lost_exp_enhlost$gene, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$lost_exp_enhlost$gene)
analyze_geneset_enrichment(genes_interest=unique(EGroupC4_intersect), genes_univ=unique(target_map$gene), pathways_list=msigdb_H_list)

clusprof_enrich_H = enricher(gene=unique(EGroupC4_intersect), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

# gsea
### Filter by expression & cell line existence & enhancer change
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts2, K27ac_dba2=K27ac_dba_deseq_contrasts2, contrasts=c(1,3,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts2, diffexprlist2=diffexpr_list_contrasts2, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }



```




```{r}
### validation statistics of gained enhancers
EGroupC4_intersect_enh1 = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain$enhancer)

EGroupC4_intersect_enh2 = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp$enhancer)

EGroupC4_intersect_enh3 = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enh$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enh$enhancer)

EGroupC4_intersect_enh4 = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain$enhancer)

print(paste0("Gain: ", length(EGroupC4_intersect_enh1)))
print(paste0("Gain & exp: ", length(EGroupC4_intersect_enh2)))
print(paste0("Gain & exp & enh: ", length(EGroupC4_intersect_enh3)))
print(paste0("Gain & exp & enh gain: ", length(EGroupC4_intersect_enh4)))

```






```{r}

#### check how many validated enhancers are oxphos genes
oxphos_genes = msigdb_H_list$HALLMARK_OXIDATIVE_PHOSPHORYLATION

# get enhancer-gene pairs
somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain$enhancergene = paste0(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain$enhancer, '|', somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene)

somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain$enhancergene = paste0(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain$enhancer, '|', somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain$gene)

EGroupC4_intersect_enhgene = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain$enhancergene, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain$enhancergene)



#### check how many validated enhancers are oxphos genes
gained_oxphos = data.frame(enhancer=unlist(lapply(strsplit(EGroupC4_intersect_enhgene, split='|', fixed=T), function(x){x[1]})),
                           gene=unlist(lapply(strsplit(EGroupC4_intersect_enhgene, split='|', fixed=T), function(x){x[2]})), 
                           enhancergene=EGroupC4_intersect_enhgene, stringsAsFactors = F)
gained_oxphos$oxphos = unlist(lapply(gained_oxphos$gene, function(x){x %in% oxphos_genes}))
gained_oxphos = gained_oxphos[gained_oxphos$oxphos,]
# add enhancer target correlation
target_map$enhancergene = paste0(target_map$enhancer, '|', target_map$gene)
gained_oxphos = merge(gained_oxphos, target_map[, c("enhancergene", "rho", "promK27ac_rho")], by="enhancergene")

print("Validated enhancer-genes assoc with oxphos")
print(gained_oxphos)
print(paste0("Num validated enhancers assoc with oxphos genes: ", length(unique(gained_oxphos$enhancer))))

### check how many validated enhancers' genes are oxphos genes
print (paste0("Num validated enhancers' genes are oxphos genes: ", length(unique(gained_oxphos$gene))))




```


```{r fig.width=4, fig.asp=1.5}
### boxplot enhancer & expression for selected genes: "COX15"  "AFG3L2" "COX10"  "NNT"    "ACO2"   "NDUFS4" "MRPS11" SLX4IP SMYD2 MAP3K21
gained_C4 = data.frame(enhancer=unlist(lapply(strsplit(EGroupC4_intersect_enhgene, split='|', fixed=T), function(x){x[1]})),
                           gene=unlist(lapply(strsplit(EGroupC4_intersect_enhgene, split='|', fixed=T), function(x){x[2]})), 
                           enhancergene=EGroupC4_intersect_enhgene, stringsAsFactors = F)
target_map$enhancergene = paste0(target_map$enhancer, '|', target_map$gene)
gained_C4 = merge(gained_C4, target_map[, c("enhancergene", "rho", "promK27ac_rho")], by="enhancergene")


for (gg in c("ACO2", "NNT", "COX15", "AFG3L2", "COX10", "NDUFS4", "MRPS11", "SLX4IP", "SMYD2", "MAP3K21")) {
  print(gg)
  zz = gained_C4[gained_C4$gene==gg, ]
  ee = zz[order(zz$rho, decreasing = T),][1, "enhancer"]
  boxplot(as.numeric(K27ac_enhancer_signalsTMM_Normal[ee,]), 
          as.numeric(K27ac_enhancer_signalsTMM_Fluke[ee,]), 
          as.numeric(K27ac_enhancer_signalsTMM_C4[ee,]), 
          as.numeric(K27ac_enhancer_signalsTMM_Mixed[ee,]), 
          at=c(1,2,3,4), notch=F, col=c("grey", "cornflowerblue", "red2","green"), main=paste0(gg, " enhancer"))
  boxplot(as.numeric(expr_TPM_data_Normal[gg,]), 
          as.numeric(expr_TPM_data_Fluke[gg,]), 
          as.numeric(expr_TPM_data_C4[gg,]), 
          as.numeric(expr_TPM_data_Mixed[gg,]), 
          at=c(1,2,3,4), notch=F, col=c("grey", "cornflowerblue", "red2","green"), main=paste0(gg, " expression"))
}
```



```{r}
### get gained oxphos enhancers info
gained_oxphos_vsFlukestats = somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain[somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain$enhancergene %in% gained_oxphos$enhancergene,]
gained_oxphos_vsFlukestats = gained_oxphos_vsFlukestats[!is.na(gained_oxphos_vsFlukestats$gene),]
gained_oxphos_vsFlukestats = gained_oxphos_vsFlukestats[order(gained_oxphos_vsFlukestats$gene, gained_oxphos_vsFlukestats$enhancer),]

gained_oxphos_vsMixedstats = somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain[somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain$enhancergene %in% gained_oxphos$enhancergene,]
gained_oxphos_vsMixedstats = gained_oxphos_vsMixedstats[!is.na(gained_oxphos_vsMixedstats$gene),]
gained_oxphos_vsMixedstats = gained_oxphos_vsMixedstats[order(gained_oxphos_vsMixedstats$gene, gained_oxphos_vsMixedstats$enhancer),]


# add enhancer-gene correlation stats
target_map$enhancergene = paste0(target_map$enhancer, '|', target_map$gene)
gained_oxphos_vsFlukestats = merge(gained_oxphos_vsFlukestats, target_map[, c("enhancergene", "rho", "promK27ac_rho")], by="enhancergene")
gained_oxphos_vsMixedstats = merge(gained_oxphos_vsMixedstats, target_map[, c("enhancergene", "rho", "promK27ac_rho")], by="enhancergene")

print(gained_oxphos_vsFlukestats)
print(gained_oxphos_vsMixedstats)

```

```{r}
# print gained enhancers in bed format to put in ucsc genome browser multiview
zz = gained_oxphos_vsFlukestats[gained_oxphos_vsFlukestats$gene=="ACO2", ]
zz = zz[order(zz$rho, decreasing = T),][1:min(5, nrow(zz)), ]
bed_coords = data.frame(chr=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){x[1]})), 
                        start=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[2])-500})),
                        end=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[3])+500})), stringsAsFactors = F)


zz = gained_oxphos_vsFlukestats[gained_oxphos_vsFlukestats$gene=="NNT", ]
zz = zz[order(zz$rho, decreasing = T),][1:min(15, nrow(zz)), ]
bed_coords = data.frame(chr=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){x[1]})), 
                        start=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[2])-500})),
                        end=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[3])+500})), stringsAsFactors = F)



zz = gained_oxphos_vsFlukestats[gained_oxphos_vsFlukestats$gene=="NDUFS4", ]
zz = zz[order(zz$rho, decreasing = T),][1:min(5, nrow(zz)), ]
bed_coords = data.frame(chr=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){x[1]})), 
                        start=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[2])-500})),
                        end=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[3])+500})), stringsAsFactors = F)


zz = gained_oxphos_vsFlukestats[gained_oxphos_vsFlukestats$gene=="COX10", ]
zz = zz[order(zz$rho, decreasing = T),][1:min(5, nrow(zz)), ]
bed_coords = data.frame(chr=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){x[1]})), 
                        start=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[2])-500})),
                        end=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[3])+500})), stringsAsFactors = F)

zz = gained_oxphos_vsFlukestats[gained_oxphos_vsFlukestats$gene=="COX15", ]
zz = zz[order(zz$rho, decreasing = T),][1:min(5, nrow(zz)), ]
bed_coords = data.frame(chr=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){x[1]})), 
                        start=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[2])-500})),
                        end=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[3])+500})), stringsAsFactors = F)


```







```{r}
### validation statistics of lost enhancers
EGroupC4_intersect_enh1 = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$lost$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$lost$enhancer)

EGroupC4_intersect_enh2 = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$lost_exp$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$lost_exp$enhancer)

EGroupC4_intersect_enh3 = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$lost_exp_enh$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$lost_exp_enh$enhancer)

EGroupC4_intersect_enh4 = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$lost_exp_enhlost$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$lost_exp_enhlost$enhancer)

print(paste0("lost: ", length(EGroupC4_intersect_enh1)))
print(paste0("lost & exp: ", length(EGroupC4_intersect_enh2)))
print(paste0("lost & exp & enh: ", length(EGroupC4_intersect_enh3)))
print(paste0("lost & exp & enh lost: ", length(EGroupC4_intersect_enh4)))
```




# EGroup Mixed somatic enhancers
```{r}

load(file="Rdata_cca_k27ac_3_cells_diffbind.Rdata")
rm(K27ac_cells_dba, K27ac_cells_enhsup_tisspeaks_dba)
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts3.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts3_diffexpr.Rdata")
```



### EGroup Mixed vs Fluke (w/ Norm ref) - just validate vs normal
```{r fig.width=8}
contrast = 5

somatic_enhancers_groupspec[[contrast]] = validate_diff_enhancers_multicomp_exp_cellline(K27ac_dba.x.1=K27ac_dba_deseq_contrasts3, K27ac_dba.x.2=K27ac_dba_deseq_contrasts3, contrasts.x=c(1, 3),
                                              group_interest.x=c(1,1), method.x=DBA_DESEQ2, diffexprlist.x.1=diffexpr_list_contrasts3, diffexprlist.x.2=diffexpr_list_contrasts3,  
                                               K27ac_dba.y.1=K27ac_cells_tisspeaks_dba, K27ac_dba.y.2=K27ac_cells_tisspeaks_dba, contrasts.y=c(3,3), group_interest.y=c(1,1),
                                               method.y=DBA_DESEQ2, diffexprlist.y.1=diffexpr_cells_list, diffexprlist.y.2=diffexpr_cells_list,
                                               corr_data=target_map, enh_present_thr=.5, filterEnh=NULL, pathways_list=msigdb_H_list, enh_p_or_q="q", exp_p_or_q="p")

names(somatic_enhancers_groupspec)[[contrast]] = "Tumor_EGroupMixed_vs_Tumor_EGroupFluke"

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$gain_exp_enhgain$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$lost_exp_enhlost$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}




### gsea
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts3, K27ac_dba2=K27ac_dba_deseq_contrasts3, contrasts=c(1,3), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts3, diffexprlist2=diffexpr_list_contrasts3, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }

```










### EGroup Mixed vs C4 (w/ Norm ref) - just validate vs normal
```{r fig.width=8}
contrast = 6

somatic_enhancers_groupspec[[contrast]] = validate_diff_enhancers_multicomp_exp_cellline(K27ac_dba.x.1=K27ac_dba_deseq_contrasts3, K27ac_dba.x.2=K27ac_dba_deseq_contrasts3, contrasts.x=c(1, 4),
                                              group_interest.x=c(1,1), method.x=DBA_DESEQ2, diffexprlist.x.1=diffexpr_list_contrasts3, diffexprlist.x.2=diffexpr_list_contrasts3,  
                                               K27ac_dba.y.1=K27ac_cells_tisspeaks_dba, K27ac_dba.y.2=K27ac_cells_tisspeaks_dba, contrasts.y=c(3,3), group_interest.y=c(1,1),
                                               method.y=DBA_DESEQ2, diffexprlist.y.1=diffexpr_cells_list, diffexprlist.y.2=diffexpr_cells_list,
                                               corr_data=target_map, enh_present_thr=.5, filterEnh=NULL, pathways_list=msigdb_H_list, enh_p_or_q="q", exp_p_or_q="p")

names(somatic_enhancers_groupspec)[[contrast]] = "Tumor_EGroupMixed_vs_Tumor_EGroupC4"

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$gain_exp_enhgain$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$lost_exp_enhlost$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}




### gsea
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts3, K27ac_dba2=K27ac_dba_deseq_contrasts3, contrasts=c(1,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts3, diffexprlist2=diffexpr_list_contrasts3, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }
```










### EGroup Mixed vs Fluke & C4

```{r fig.width=8}
# gain
EGroupMixed_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain$gene)

analyze_geneset_enrichment(genes_interest=unique(EGroupMixed_intersect), genes_univ=unique(target_map$gene), pathways_list=msigdb_H_list)

clusprof_enrich_H = enricher(gene=unique(EGroupMixed_intersect), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}




# lost
EGroupMixed_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$lost_exp_enhlost$gene, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$lost_exp_enhlost$gene)

analyze_geneset_enrichment(genes_interest=unique(EGroupMixed_intersect), genes_univ=unique(target_map$gene), pathways_list=msigdb_H_list)

clusprof_enrich_H = enricher(gene=unique(EGroupMixed_intersect), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}




# gsea
### Filter by expression & cell line existence & enhancer change
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts3, K27ac_dba2=K27ac_dba_deseq_contrasts3, contrasts=c(1,3,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts3, diffexprlist2=diffexpr_list_contrasts3, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}


if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }



```






```{r}
### validation statistics of gained enhancers
EGroupMixed_intersect_enh1 = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain$enhancer)

EGroupMixed_intersect_enh2 = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp$enhancer)

EGroupMixed_intersect_enh3 = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enh$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enh$enhancer)

EGroupMixed_intersect_enh4 = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain$enhancer)

print(paste0("Gain: ", length(EGroupMixed_intersect_enh1)))
print(paste0("Gain & exp: ", length(EGroupMixed_intersect_enh2)))
print(paste0("Gain & exp & enh: ", length(EGroupMixed_intersect_enh3)))
print(paste0("Gain & exp & enh gain: ", length(EGroupMixed_intersect_enh4)))

```

```{r}

#### check how many validated enhancers are immune genes
ig_genes = msigdb_H_list$HALLMARK_INTERFERON_GAMMA_RESPONSE
ia_genes = msigdb_H_list$HALLMARK_INTERFERON_ALPHA_RESPONSE
il2_genes = msigdb_H_list$HALLMARK_IL2_STAT5_SIGNALING
il6_genes = msigdb_H_list$HALLMARK_IL6_JAK_STAT3_SIGNALING


# get enhancer-gene pairs
somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain$enhancergene = paste0(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain$enhancer, '|', somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene)

somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain$enhancergene = paste0(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain$enhancer, '|', somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain$gene)

EGroupMixed_intersect_enhgene = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain$enhancergene, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain$enhancergene)


#### check how many validated enhancers are immune genes
gained_imm = data.frame(enhancer=unlist(lapply(strsplit(EGroupMixed_intersect_enhgene, split='|', fixed=T), function(x){x[1]})),
                        gene=unlist(lapply(strsplit(EGroupMixed_intersect_enhgene, split='|', fixed=T), function(x){x[2]})), 
                        enhancergene=EGroupMixed_intersect_enhgene, stringsAsFactors = F)
gained_imm$ig = unlist(lapply(gained_imm$gene, function(x){x %in% ig_genes}))
gained_imm$ia = unlist(lapply(gained_imm$gene, function(x){x %in% ia_genes}))
gained_imm$il2 = unlist(lapply(gained_imm$gene, function(x){x %in% il2_genes}))
gained_imm$il6 = unlist(lapply(gained_imm$gene, function(x){x %in% il6_genes}))
gained_imm = gained_imm[gained_imm$ig | gained_imm$ia | gained_imm$il2 | gained_imm$il6,]
# add enhancer target correlation
target_map$enhancergene = paste0(target_map$enhancer, '|', target_map$gene)
gained_imm = merge(gained_imm, target_map[, c("enhancergene", "rho", "promK27ac_rho")], by="enhancergene")

print("Validated enhancer-genes assoc with immune")
print(gained_imm)
print(paste0("Num validated enhancers assoc with int gamma genes: ", length(unique(gained_imm$enhancer[gained_imm$ig]))))
print(paste0("Num validated enhancers assoc with int alpha genes: ", length(unique(gained_imm$enhancer[gained_imm$ia]))))
print(paste0("Num validated enhancers assoc with il2 genes: ", length(unique(gained_imm$enhancer[gained_imm$il2]))))
print(paste0("Num validated enhancers assoc with il6 genes: ", length(unique(gained_imm$enhancer[gained_imm$il6]))))

### check how many validated enhancers' genes are immune genes
print (paste0("Num validated enhancers' genes are int gamma genes: ", length(unique(gained_imm$gene[gained_imm$ig]))))
print (paste0("Num validated enhancers' genes are int alpha genes: ", length(unique(gained_imm$gene[gained_imm$ia]))))
print (paste0("Num validated enhancers' genes are il2 genes: ", length(unique(gained_imm$gene[gained_imm$il2]))))
print (paste0("Num validated enhancers' genes are il6 genes: ", length(unique(gained_imm$gene[gained_imm$il6]))))




```




```{r fig.width=4, fig.asp=1.5}

gained_Mixed = data.frame(enhancer=unlist(lapply(strsplit(EGroupMixed_intersect_enhgene, split='|', fixed=T), function(x){x[1]})),
                gene=unlist(lapply(strsplit(EGroupMixed_intersect_enhgene, split='|', fixed=T), function(x){x[2]})), 
                enhancergene=EGroupMixed_intersect_enhgene, stringsAsFactors = F)
target_map$enhancergene = paste0(target_map$enhancer, '|', target_map$gene)
gained_Mixed = merge(gained_Mixed, target_map[, c("enhancergene", "rho", "promK27ac_rho")], by="enhancergene")



### boxplot enhancer & expression for selected genes: CTLA4 STAT1 PARP14 CXCL9 APOL6 FCGR1A GBP4 PSME1 BST2 IFNAR2
for (gg in c("CTLA4", "STAT1", "PARP14", "CXCL9", "APOL6", "FCGR1A", "GBP4", "PSME1", "BST2", "IFNAR2")) {
  print(gg)
  zz = gained_Mixed[gained_Mixed$gene==gg, ]
  ee = zz[order(zz$rho, decreasing = T),][1, "enhancer"]
  boxplot(as.numeric(K27ac_enhancer_signalsTMM_Normal[ee,]), 
          as.numeric(K27ac_enhancer_signalsTMM_Fluke[ee,]), 
          as.numeric(K27ac_enhancer_signalsTMM_C4[ee,]), 
          as.numeric(K27ac_enhancer_signalsTMM_Mixed[ee,]), 
          at=c(1,2,3,4), notch=F, col=c("grey", "cornflowerblue", "red2","green"), main=paste0(gg, " enhancer"))
  boxplot(as.numeric(expr_TPM_data_Normal[gg,]), 
          as.numeric(expr_TPM_data_Fluke[gg,]), 
          as.numeric(expr_TPM_data_C4[gg,]), 
          as.numeric(expr_TPM_data_Mixed[gg,]), 
          at=c(1,2,3,4), notch=F, col=c("grey", "cornflowerblue", "red2","green"), main=paste0(gg, " expression"))
}
```


```{r}
### get gained imm enhancers info
gained_imm_vsFlukestats = somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain[somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain$enhancergene %in% gained_imm$enhancergene,]
gained_imm_vsFlukestats = gained_imm_vsFlukestats[!is.na(gained_imm_vsFlukestats$gene),]
gained_imm_vsFlukestats = gained_imm_vsFlukestats[order(gained_imm_vsFlukestats$gene, gained_imm_vsFlukestats$enhancer),]

gained_imm_vsC4stats = somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain[somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain$enhancergene %in% gained_imm$enhancergene,]
gained_imm_vsC4stats = gained_imm_vsC4stats[!is.na(gained_imm_vsC4stats$gene),]
gained_imm_vsC4stats = gained_imm_vsC4stats[order(gained_imm_vsC4stats$gene, gained_imm_vsC4stats$enhancer),]


# add enhancer-gene correlation stats
target_map$enhancergene = paste0(target_map$enhancer, '|', target_map$gene)
gained_imm_vsFlukestats = merge(gained_imm_vsFlukestats, target_map[, c("enhancergene", "rho", "promK27ac_rho")], by="enhancergene")
gained_imm_vsC4stats = merge(gained_imm_vsC4stats, target_map[, c("enhancergene", "rho", "promK27ac_rho")], by="enhancergene")

print(gained_imm_vsFlukestats)
print(gained_imm_vsC4stats)

```

```{r}
# print gained enhancers in bed format to put in ucsc genome browser multiview
zz = gained_imm_vsFlukestats[gained_imm_vsFlukestats$gene=="CTLA4", ]
zz = zz[order(zz$rho, decreasing = T),][1:min(5, nrow(zz)), ]
bed_coords = data.frame(chr=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){x[1]})), 
                        start=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[2])-500})),
                        end=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[3])+500})), stringsAsFactors = F)


zz = gained_imm_vsFlukestats[gained_imm_vsFlukestats$gene=="STAT1", ]
zz = zz[order(zz$rho, decreasing = T),][1:min(5, nrow(zz)), ]
bed_coords = data.frame(chr=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){x[1]})), 
                        start=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[2])-500})),
                        end=unlist(lapply(strsplit(zz$enhancer, "[:-]"), function(x){strtoi(x[3])+500})), stringsAsFactors = F)


```


```{r}
### validation statistics of losted enhancers
EGroupMixed_intersect_enh1 = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$lost$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$lost$enhancer)

EGroupMixed_intersect_enh2 = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$lost_exp$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$lost_exp$enhancer)

EGroupMixed_intersect_enh3 = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$lost_exp_enh$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$lost_exp_enh$enhancer)

EGroupMixed_intersect_enh4 = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$lost_exp_enhlost$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$lost_exp_enhlost$enhancer)

print(paste0("lost: ", length(EGroupMixed_intersect_enh1)))
print(paste0("lost & exp: ", length(EGroupMixed_intersect_enh2)))
print(paste0("lost & exp & enh: ", length(EGroupMixed_intersect_enh3)))
print(paste0("lost & exp & enh lost: ", length(EGroupMixed_intersect_enh4)))

```





# EGroup Mixed-Clus3 somatic enhancers
```{r}

load(file="Rdata_cca_k27ac_3_cells_diffbind.Rdata")
rm(K27ac_cells_dba, K27ac_cells_enhsup_tisspeaks_dba)
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts4.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts4_diffexpr.Rdata")
```



### EGroup Mixed-Clus3 vs Fluke (w/ Norm ref) - just validate vs normal
```{r fig.width=8}
contrast = 7

somatic_enhancers_groupspec[[contrast]] = validate_diff_enhancers_multicomp_exp_cellline(K27ac_dba.x.1=K27ac_dba_deseq_contrasts4, K27ac_dba.x.2=K27ac_dba_deseq_contrasts4, contrasts.x=c(1, 3),
                                              group_interest.x=c(1,1), method.x=DBA_DESEQ2, diffexprlist.x.1=diffexpr_list_contrasts4, diffexprlist.x.2=diffexpr_list_contrasts4,  
                                               K27ac_dba.y.1=K27ac_cells_tisspeaks_dba, K27ac_dba.y.2=K27ac_cells_tisspeaks_dba, contrasts.y=c(3,3), group_interest.y=c(1,1),
                                               method.y=DBA_DESEQ2, diffexprlist.y.1=diffexpr_cells_list, diffexprlist.y.2=diffexpr_cells_list,
                                               corr_data=target_map, enh_present_thr=.5, filterEnh=NULL, pathways_list=msigdb_H_list, enh_p_or_q="q", exp_p_or_q="p")

names(somatic_enhancers_groupspec)[[contrast]] = "Tumor_EGroupMixedClus3_vs_Tumor_EGroupFluke"

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$gain_exp_enhgain$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$lost_exp_enhlost$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}



### gsea
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts4, K27ac_dba2=K27ac_dba_deseq_contrasts4, contrasts=c(1,3), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts4, diffexprlist2=diffexpr_list_contrasts4, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }
```





### EGroup Mixed-Clus3 vs C4 (w/ Norm ref) - just validate vs normal
```{r fig.width=8}
contrast = 8

somatic_enhancers_groupspec[[contrast]] = validate_diff_enhancers_multicomp_exp_cellline(K27ac_dba.x.1=K27ac_dba_deseq_contrasts4, K27ac_dba.x.2=K27ac_dba_deseq_contrasts4, contrasts.x=c(1, 4),
                                              group_interest.x=c(1,1), method.x=DBA_DESEQ2, diffexprlist.x.1=diffexpr_list_contrasts4, diffexprlist.x.2=diffexpr_list_contrasts4,  
                                               K27ac_dba.y.1=K27ac_cells_tisspeaks_dba, K27ac_dba.y.2=K27ac_cells_tisspeaks_dba, contrasts.y=c(3,3), group_interest.y=c(1,1),
                                               method.y=DBA_DESEQ2, diffexprlist.y.1=diffexpr_cells_list, diffexprlist.y.2=diffexpr_cells_list,
                                               corr_data=target_map, enh_present_thr=.5, filterEnh=NULL, pathways_list=msigdb_H_list, enh_p_or_q='q', exp_p_or_q="p")

names(somatic_enhancers_groupspec)[[contrast]] = "Tumor_EGroupMixedClus3_vs_Tumor_EGroupC4"

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$gain_exp_enhgain$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$lost_exp_enhlost$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}



### gsea
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts4, K27ac_dba2=K27ac_dba_deseq_contrasts4, contrasts=c(1,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts4, diffexprlist2=diffexpr_list_contrasts4, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }
```







### EGroup Mixed-Clus3 vs Fluke & C4

```{r fig.width=8}

# gain
EGroupMixedClus3_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixedClus3_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene,
                                         somatic_enhancers_groupspec$Tumor_EGroupMixedClus3_vs_Tumor_EGroupC4$gain_exp_enhgain$gene)
analyze_geneset_enrichment(genes_interest=unique(EGroupMixedClus3_intersect), genes_univ=unique(target_map$gene), pathways_list=msigdb_H_list)

clusprof_enrich_H = enricher(gene=unique(EGroupMixedClus3_intersect), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

# lost
EGroupMixedClus3_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixedClus3_vs_Tumor_EGroupFluke$lost_exp_enhlost$gene,
                                         somatic_enhancers_groupspec$Tumor_EGroupMixedClus3_vs_Tumor_EGroupC4$lost_exp_enhlost$gene)
analyze_geneset_enrichment(genes_interest=unique(EGroupMixedClus3_intersect), genes_univ=unique(target_map$gene), pathways_list=msigdb_H_list)

clusprof_enrich_H = enricher(gene=unique(EGroupMixedClus3_intersect), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}



# gsea
### Filter by expression & cell line existence & enhancer change
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts4, K27ac_dba2=K27ac_dba_deseq_contrasts4, contrasts=c(1,3,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts4, diffexprlist2=diffexpr_list_contrasts4, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }



```






# EGroup Mixed-AA somatic enhancers
```{r}

load(file="Rdata_cca_k27ac_3_cells_diffbind.Rdata")
rm(K27ac_cells_dba, K27ac_cells_enhsup_tisspeaks_dba)
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts5.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts5_diffexpr.Rdata")
```


### EGroup Mixed-AA vs Fluke (w/ Norm ref) - just validate vs normal
```{r fig.width=8}
contrast = 9

somatic_enhancers_groupspec[[contrast]] = validate_diff_enhancers_multicomp_exp_cellline(K27ac_dba.x.1=K27ac_dba_deseq_contrasts5, K27ac_dba.x.2=K27ac_dba_deseq_contrasts5, contrasts.x=c(1, 3),
                                              group_interest.x=c(1,1), method.x=DBA_DESEQ2, diffexprlist.x.1=diffexpr_list_contrasts5, diffexprlist.x.2=diffexpr_list_contrasts5,  
                                               K27ac_dba.y.1=K27ac_cells_tisspeaks_dba, K27ac_dba.y.2=K27ac_cells_tisspeaks_dba, contrasts.y=c(3,3), group_interest.y=c(1,1),
                                               method.y=DBA_DESEQ2, diffexprlist.y.1=diffexpr_cells_list, diffexprlist.y.2=diffexpr_cells_list,
                                               corr_data=target_map, enh_present_thr=.5, filterEnh=NULL, pathways_list=msigdb_H_list, enh_p_or_q="q", exp_p_or_q="p")

names(somatic_enhancers_groupspec)[[contrast]] = "Tumor_EGroupMixedAA_vs_Tumor_EGroupFluke"

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$gain_exp_enhgain$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$lost_exp_enhlost$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}


### gsea
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts5, K27ac_dba2=K27ac_dba_deseq_contrasts5, contrasts=c(1,3), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts5, diffexprlist2=diffexpr_list_contrasts5, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }
```




### EGroup Mixed-AA vs C4 (w/ Norm ref) - just validate vs normal
```{r fig.width=8}
contrast = 10

somatic_enhancers_groupspec[[contrast]] = validate_diff_enhancers_multicomp_exp_cellline(K27ac_dba.x.1=K27ac_dba_deseq_contrasts5, K27ac_dba.x.2=K27ac_dba_deseq_contrasts5, contrasts.x=c(1, 4),
                                              group_interest.x=c(1,1), method.x=DBA_DESEQ2, diffexprlist.x.1=diffexpr_list_contrasts5, diffexprlist.x.2=diffexpr_list_contrasts5,  
                                               K27ac_dba.y.1=K27ac_cells_tisspeaks_dba, K27ac_dba.y.2=K27ac_cells_tisspeaks_dba, contrasts.y=c(3,3), group_interest.y=c(1,1),
                                               method.y=DBA_DESEQ2, diffexprlist.y.1=diffexpr_cells_list, diffexprlist.y.2=diffexpr_cells_list,
                                               corr_data=target_map, enh_present_thr=.5, filterEnh=NULL, pathways_list=msigdb_H_list, enh_p_or_q="q", exp_p_or_q="p")
names(somatic_enhancers_groupspec)[[contrast]] = "Tumor_EGroupMixedAA_vs_Tumor_EGroupC4"

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$gain_exp_enhgain$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$lost_exp_enhlost$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

### gsea
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts5, K27ac_dba2=K27ac_dba_deseq_contrasts5, contrasts=c(1,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts5, diffexprlist2=diffexpr_list_contrasts5, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }
```





### use intersection of EGroup Mixed-AA vs Fluke, C4 (ie. exclude Mixed-NonAA!)

```{r fig.width=8}

# gain
EGroupMixedAA_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixedAA_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene,
                                         somatic_enhancers_groupspec$Tumor_EGroupMixedAA_vs_Tumor_EGroupC4$gain_exp_enhgain$gene)
analyze_geneset_enrichment(genes_interest=unique(EGroupMixedAA_intersect), genes_univ=unique(target_map$gene), pathways_list=msigdb_H_list)

clusprof_enrich_H = enricher(gene=unique(EGroupMixedAA_intersect), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}


# lost
EGroupMixedAA_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixedAA_vs_Tumor_EGroupFluke$lost_exp_enhlost$gene,
                                         somatic_enhancers_groupspec$Tumor_EGroupMixedAA_vs_Tumor_EGroupC4$lost_exp_enhlost$gene)
analyze_geneset_enrichment(genes_interest=unique(EGroupMixedAA_intersect), genes_univ=unique(target_map$gene), pathways_list=msigdb_H_list)

clusprof_enrich_H = enricher(gene=unique(EGroupMixedAA_intersect), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}


# gsea
### Filter by expression & cell line existence & enhancer change
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts5, K27ac_dba2=K27ac_dba_deseq_contrasts5, contrasts=c(1,3,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts5, diffexprlist2=diffexpr_list_contrasts5, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }



```









# EGroup Mixed-nonAAC3 somatic enhancers
```{r}

load(file="Rdata_cca_k27ac_3_cells_diffbind.Rdata")
rm(K27ac_cells_dba, K27ac_cells_enhsup_tisspeaks_dba)
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts6.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts6_diffexpr.Rdata")
```


### EGroup Mixed-nonAAC3 vs Fluke (w/ Norm ref) - just validate vs normal
```{r fig.width=8}
contrast = 11

somatic_enhancers_groupspec[[contrast]] = validate_diff_enhancers_multicomp_exp_cellline(K27ac_dba.x.1=K27ac_dba_deseq_contrasts6, K27ac_dba.x.2=K27ac_dba_deseq_contrasts6, contrasts.x=c(1, 3),
                                              group_interest.x=c(1,1), method.x=DBA_DESEQ2, diffexprlist.x.1=diffexpr_list_contrasts6, diffexprlist.x.2=diffexpr_list_contrasts6,  
                                               K27ac_dba.y.1=K27ac_cells_tisspeaks_dba, K27ac_dba.y.2=K27ac_cells_tisspeaks_dba, contrasts.y=c(3,3), group_interest.y=c(1,1),
                                               method.y=DBA_DESEQ2, diffexprlist.y.1=diffexpr_cells_list, diffexprlist.y.2=diffexpr_cells_list,
                                               corr_data=target_map, enh_present_thr=.5, filterEnh=NULL, pathways_list=msigdb_H_list, enh_p_or_q="q", exp_p_or_q="p")

names(somatic_enhancers_groupspec)[[contrast]] = "Tumor_EGroupMixedNonAAC3_vs_Tumor_EGroupFluke"

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$gain_exp_enhgain$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$lost_exp_enhlost$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}


### gsea
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts6, K27ac_dba2=K27ac_dba_deseq_contrasts6, contrasts=c(1,3), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts6, diffexprlist2=diffexpr_list_contrasts6, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }
```




### EGroup Mixed-nonAAC3 vs C4 (w/ Norm ref) - just validate vs normal
```{r fig.width=8}
contrast = 12

somatic_enhancers_groupspec[[contrast]] = validate_diff_enhancers_multicomp_exp_cellline(K27ac_dba.x.1=K27ac_dba_deseq_contrasts6, K27ac_dba.x.2=K27ac_dba_deseq_contrasts6, contrasts.x=c(1, 4),
                                              group_interest.x=c(1,1), method.x=DBA_DESEQ2, diffexprlist.x.1=diffexpr_list_contrasts6, diffexprlist.x.2=diffexpr_list_contrasts6,  
                                               K27ac_dba.y.1=K27ac_cells_tisspeaks_dba, K27ac_dba.y.2=K27ac_cells_tisspeaks_dba, contrasts.y=c(3,3), group_interest.y=c(1,1),
                                               method.y=DBA_DESEQ2, diffexprlist.y.1=diffexpr_cells_list, diffexprlist.y.2=diffexpr_cells_list,
                                               corr_data=target_map, enh_present_thr=.5, filterEnh=NULL, pathways_list=msigdb_H_list, enh_p_or_q="q", exp_p_or_q="p")
names(somatic_enhancers_groupspec)[[contrast]] = "Tumor_EGroupMixedNonAAC3_vs_Tumor_EGroupC4"

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$gain_exp_enhgain$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

clusprof_enrich_H = enricher(gene=unique(somatic_enhancers_groupspec[[contrast]]$lost_exp_enhlost$gene), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}

### gsea
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts6, K27ac_dba2=K27ac_dba_deseq_contrasts6, contrasts=c(1,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts6, diffexprlist2=diffexpr_list_contrasts6, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }
```





### EGroup Mixed-nonAAC3 vs Fluke, C4 

```{r fig.width=8}

# gain
EGroupMixedNonAAC3_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixedNonAAC3_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene,
                                         somatic_enhancers_groupspec$Tumor_EGroupMixedNonAAC3_vs_Tumor_EGroupC4$gain_exp_enhgain$gene)
analyze_geneset_enrichment(genes_interest=unique(EGroupMixedNonAAC3_intersect), genes_univ=unique(target_map$gene), pathways_list=msigdb_H_list)

clusprof_enrich_H = enricher(gene=unique(EGroupMixedNonAAC3_intersect), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}


# lost
EGroupMixedNonAAC3_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixedNonAAC3_vs_Tumor_EGroupFluke$lost_exp_enhlost$gene,
                                         somatic_enhancers_groupspec$Tumor_EGroupMixedNonAAC3_vs_Tumor_EGroupC4$lost_exp_enhlost$gene)
analyze_geneset_enrichment(genes_interest=unique(EGroupMixedNonAAC3_intersect), genes_univ=unique(target_map$gene), pathways_list=msigdb_H_list)

clusprof_enrich_H = enricher(gene=unique(EGroupMixedNonAAC3_intersect), universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)
if (!is.null(clusprof_enrich_H)) { dotplot(clusprof_enrich_H, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (Enrich)", font.size=10) }
if (!is.null(clusprof_enrich_H)) { 
  zz = attributes(clusprof_enrich_H)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}


# gsea
### Filter by expression & cell line existence & enhancer change
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts6, K27ac_dba2=K27ac_dba_deseq_contrasts6, contrasts=c(1,3,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts6, diffexprlist2=diffexpr_list_contrasts6, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)
print(paste0("Num 0's: ", sum(gsea_genelist_CellFilt2==0)))

clusprof_gsea_H = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000)

clusprof_gsea1 = clusprof_gsea_H
attributes(clusprof_gsea1)$result = attributes(clusprof_gsea1)$result[attributes(clusprof_gsea1)$result$enrichmentScore>0,]
print(paste0("Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA): ", nrow(clusprof_gsea1)))
if (nrow(clusprof_gsea1)>0) { dotplot(clusprof_gsea1, showCategory = 50, title = "Filter expr + cell existence + enh change, Group1 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea1)>0) { 
  zz = attributes(clusprof_gsea1)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea1)>0) { cnetplot(clusprof_gsea1, showCategory = 10, categorySize="geneNum", font.size=7) }

clusprof_gsea2 = clusprof_gsea_H
attributes(clusprof_gsea2)$result = attributes(clusprof_gsea2)$result[attributes(clusprof_gsea2)$result$enrichmentScore<0,]
print(paste0("Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA): ", nrow(clusprof_gsea2)))
if (nrow(clusprof_gsea2)>0) { dotplot(clusprof_gsea2, showCategory = 50, title = "Filter expr + cell existence + enh change, Group2 higher, Hallmark (GSEA)", font.size=10) }
if (nrow(clusprof_gsea2)>0) { 
  zz = attributes(clusprof_gsea2)$result[, c("ID", "p.adjust")]
  zz = zz[zz$p.adjust < .05,]
  zz = zz[order(zz$p.adjust),]
  zz$logFDR = -log10(zz$p.adjust)
  zz$ID = factor(zz$ID, levels = rev(unique(zz$ID)))
  ggplot(zz, aes(x=ID, y= logFDR)) +
    geom_col(fill="#e67050", alpha=.6) +
    geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
    theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))
}
if (nrow(clusprof_gsea2)>0) { cnetplot(clusprof_gsea2, showCategory = 10, categorySize="geneNum", font.size=7) }



```






```{r include=F}
# Takes a lot of time, so save it
save(somatic_enhancers_groupspec, file="Rdata_cca_k27ac_3b_somatic_enhancers_groupspec.Rdata")

# # just load it
# load("Rdata_cca_k27ac_3b_somatic_enhancers_groupspec.Rdata")
```



# plot combined pathways' FDR-barplots for publication
```{r}


load(file="Rdata_cca_k27ac_3_cells_diffbind.Rdata")
rm(K27ac_cells_dba, K27ac_cells_enhsup_tisspeaks_dba)
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts1.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts2.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts3.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts4.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts5.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts1_diffexpr.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts2_diffexpr.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts3_diffexpr.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts4_diffexpr.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts5_diffexpr.Rdata")



############ get Enrich results
get_enrich_result = function(genes, universe, TERM2GENE) {
  enrich_result = enricher(gene=genes, universe=universe, TERM2GENE=TERM2GENE)
  if (is.null(enrich_result)) { 
    return(data.frame(ID=character(), p.adjust=double(), logFDR=double(), rowID=integer()))
  }
  enrich_result = attributes(enrich_result)$result[, c("ID", "p.adjust")]
  enrich_result = enrich_result[enrich_result$p.adjust < .05,]
  if (nrow(enrich_result)==0) {
    return(data.frame(ID=character(), p.adjust=double(), logFDR=double(), rowID=integer()))
  }
  enrich_result = enrich_result[order(enrich_result$p.adjust),]
  enrich_result$logFDR = -log10(enrich_result$p.adjust)
  enrich_result$rowID = -1 : -nrow(enrich_result)
  return (enrich_result)
}

### Fluke vs C4 + Mixed
# gain
genes_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain$gene, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain$gene)
fluke_gain_enrich = get_enrich_result(genes=genes_intersect, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)

# lost
genes_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$lost_exp_enhlost$gene, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$lost_exp_enhlost$gene)
fluke_lost_enrich = get_enrich_result(genes=genes_intersect, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)

### C4 vs Fluke + Mixed
# gain
genes_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain$gene)
C4_gain_enrich = get_enrich_result(genes=genes_intersect, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)

# lost
genes_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$lost_exp_enhlost$gene, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$lost_exp_enhlost$gene)
C4_lost_enrich = get_enrich_result(genes=genes_intersect, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)

### Mixed vs Fluke + C4
# gain
genes_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain$gene)
mixed_gain_enrich = get_enrich_result(genes=genes_intersect, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)

# lost
genes_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$lost_exp_enhlost$gene, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$lost_exp_enhlost$gene)
mixed_lost_enrich = get_enrich_result(genes=genes_intersect, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)

### Mixed-Clus3 vs Fluke + C4
# gain
genes_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixedClus3_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene,
                                         somatic_enhancers_groupspec$Tumor_EGroupMixedClus3_vs_Tumor_EGroupC4$gain_exp_enhgain$gene)
mixedClus3_gain_enrich = get_enrich_result(genes=genes_intersect, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)

# lost
genes_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixedClus3_vs_Tumor_EGroupFluke$lost_exp_enhlost$gene,
                                         somatic_enhancers_groupspec$Tumor_EGroupMixedClus3_vs_Tumor_EGroupC4$lost_exp_enhlost$gene)
mixedClus3_lost_enrich = get_enrich_result(genes=genes_intersect, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)


### Mixed-AA vs Fluke + C4
# gain
genes_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixedAA_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene,
                                         somatic_enhancers_groupspec$Tumor_EGroupMixedAA_vs_Tumor_EGroupC4$gain_exp_enhgain$gene)
mixedAA_gain_enrich = get_enrich_result(genes=genes_intersect, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)

# lost
genes_intersect = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixedAA_vs_Tumor_EGroupFluke$lost_exp_enhlost$gene,
                                         somatic_enhancers_groupspec$Tumor_EGroupMixedAA_vs_Tumor_EGroupC4$lost_exp_enhlost$gene)
mixedAA_lost_enrich = get_enrich_result(genes=genes_intersect, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene)


############ get gsea results
get_gsea_result = function(genelist, TERM2GENE) {
  gsea_result = GSEA(geneList=genelist, minGSSize=1, maxGSSize=5000, TERM2GENE=TERM2GENE, seed=12345, nPerm=1000)
  
  gsea_result_gain = gsea_result
  attributes(gsea_result_gain)$result = attributes(gsea_result_gain)$result[attributes(gsea_result_gain)$result$enrichmentScore>0,]
  if (nrow(gsea_result_gain)>0) { 
    gsea_result_gain = attributes(gsea_result_gain)$result[, c("ID", "p.adjust")]
    gsea_result_gain = gsea_result_gain[gsea_result_gain$p.adjust < .05,]
    gsea_result_gain = gsea_result_gain[order(gsea_result_gain$p.adjust),]
    gsea_result_gain$logFDR = -log10(gsea_result_gain$p.adjust)
    gsea_result_gain$rowID = -1 : -nrow(gsea_result_gain)
  }
  else { gsea_result_gain = data.frame(ID=character(), p.adjust=double(), logFDR=double(), rowID=integer()) }
  
  gsea_result_lost = gsea_result
  attributes(gsea_result_lost)$result = attributes(gsea_result_lost)$result[attributes(gsea_result_lost)$result$enrichmentScore<0,]
  if (nrow(gsea_result_lost)>0) { 
    gsea_result_lost = attributes(gsea_result_lost)$result[, c("ID", "p.adjust")]
    gsea_result_lost = gsea_result_lost[gsea_result_lost$p.adjust < .05,]
    gsea_result_lost = gsea_result_lost[order(gsea_result_lost$p.adjust),]
    gsea_result_lost$logFDR = -log10(gsea_result_lost$p.adjust)
    gsea_result_lost$rowID = -1 : -nrow(gsea_result_lost)
  }
  else { gsea_result_lost = data.frame(ID=character(), p.adjust=double(), logFDR=double(), rowID=integer()) }
  return (list("gain"=gsea_result_gain,"lost"=gsea_result_lost))
}

### Fluke vs C4 + Mixed
genelist = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts1, K27ac_dba2=K27ac_dba_deseq_contrasts1, contrasts=c(1,3,4),
                                                       group_interest=c(1,1), method=DBA_DESEQ2,   
                                                       diffexprlist1=diffexpr_list_contrasts1, diffexprlist2=diffexpr_list_contrasts1, corr_data=target_map,
                                                       filterEnh=NULL, filterExpr=T,  filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=2,
                                                       group_interest.cell=1, enh_present_thr=.5)
genelist = sort(genelist, decreasing = T)
gsea_res = get_gsea_result(genelist, TERM2GENE=msigdb_H_term2gene)
fluke_gain_gsea = gsea_res$gain
fluke_lost_gsea = gsea_res$lost

### C4 vs Fluke + Mixed
genelist = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts2, K27ac_dba2=K27ac_dba_deseq_contrasts2, contrasts=c(1,3,4),
                                                       group_interest=c(1,1), method=DBA_DESEQ2,   
                                                       diffexprlist1=diffexpr_list_contrasts2, diffexprlist2=diffexpr_list_contrasts2, corr_data=target_map,
                                                       filterEnh=NULL, filterExpr=T,  filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3,
                                                       group_interest.cell=1, enh_present_thr=.5)
genelist = sort(genelist, decreasing = T)
gsea_res = get_gsea_result(genelist, TERM2GENE=msigdb_H_term2gene)
C4_gain_gsea = gsea_res$gain
C4_lost_gsea = gsea_res$lost

### Mixed vs Fluke + C4
genelist = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts3, K27ac_dba2=K27ac_dba_deseq_contrasts3, contrasts=c(1,3,4),
                                                       group_interest=c(1,1), method=DBA_DESEQ2,   
                                                       diffexprlist1=diffexpr_list_contrasts3, diffexprlist2=diffexpr_list_contrasts3, corr_data=target_map,
                                                       filterEnh=NULL, filterExpr=T,  filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3,
                                                       group_interest.cell=1, enh_present_thr=.5)
genelist = sort(genelist, decreasing = T)
gsea_res = get_gsea_result(genelist, TERM2GENE=msigdb_H_term2gene)
mixed_gain_gsea = gsea_res$gain
mixed_lost_gsea = gsea_res$lost



### Mixed-Clus3 vs Fluke + C4
genelist = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts3, K27ac_dba2=K27ac_dba_deseq_contrasts3, contrasts=c(1,3,4),
                                                       group_interest=c(1,1), method=DBA_DESEQ2,   
                                                       diffexprlist1=diffexpr_list_contrasts3, diffexprlist2=diffexpr_list_contrasts3,
                                                       corr_data=target_map,
                                                       filterEnh=NULL, filterExpr=T,  filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba,
                                                       contrast.cell=3,
                                                       group_interest.cell=1, enh_present_thr=.5)
genelist = sort(genelist, decreasing = T)
gsea_res = get_gsea_result(genelist, TERM2GENE=msigdb_H_term2gene)
mixedClus3_gain_gsea = gsea_res$gain
mixedClus3_lost_gsea = gsea_res$lost



### Mixed-AA vs Fluke + C4
genelist = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts5, K27ac_dba2=K27ac_dba_deseq_contrasts5, contrasts=c(1,3,4),
                                                       group_interest=c(1,1),method=DBA_DESEQ2,           
                                                       diffexprlist1=diffexpr_list_contrasts5, diffexprlist2=diffexpr_list_contrasts5,
                                                       corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1,
                                            enh_present_thr=.5)
genelist = sort(genelist, decreasing = T)
gsea_res = get_gsea_result(genelist, TERM2GENE=msigdb_H_term2gene)
mixedAA_gain_gsea = gsea_res$gain
mixedAA_lost_gsea = gsea_res$lost
```





```{r}
####### plot combined enrich results
# gain
if (nrow(fluke_gain_enrich)>0) {
  fluke_gain_enrich$rowID = -1 : - nrow(fluke_gain_enrich)
}
if (nrow(C4_gain_enrich)>0) {
  C4_gain_enrich$rowID = -1 : - nrow(C4_gain_enrich)
  C4_gain_enrich$rowID = C4_gain_enrich$rowID - nrow(fluke_gain_enrich) - 1
}
if (nrow(mixed_gain_enrich)>0) {
  mixed_gain_enrich$rowID = -1 : - nrow(mixed_gain_enrich)
  mixed_gain_enrich$rowID = mixed_gain_enrich$rowID - nrow(fluke_gain_enrich) - 1 - nrow(C4_gain_enrich) - 1 
}
combine_gain_enrich = rbind(fluke_gain_enrich, C4_gain_enrich, mixed_gain_enrich)
ggplot(combine_gain_enrich, aes(x=rowID, y= logFDR)) +
  geom_col(fill="#e67050", alpha=.6) +
  geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
  geom_vline(xintercept = c(-nrow(fluke_gain_enrich)-1, -nrow(fluke_gain_enrich)-nrow(C4_gain_enrich)-2)) + 
  scale_x_continuous(breaks=combine_gain_enrich$rowID, labels=combine_gain_enrich$ID) +
  theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))


# lost
if (nrow(fluke_lost_enrich)>0) {
  fluke_lost_enrich$rowID = -1 : - nrow(fluke_lost_enrich)
}
if (nrow(C4_lost_enrich)>0) {
  C4_lost_enrich$rowID = -1 : - nrow(C4_lost_enrich)
  C4_lost_enrich$rowID = C4_lost_enrich$rowID - nrow(fluke_lost_enrich) - 1
}
if (nrow(mixed_lost_enrich)>0) {
  mixed_lost_enrich$rowID = -1 : - nrow(mixed_lost_enrich)
  mixed_lost_enrich$rowID = mixed_lost_enrich$rowID - nrow(fluke_lost_enrich) - 1 - nrow(C4_lost_enrich) - 1 
}
combine_lost_enrich = rbind(fluke_lost_enrich, C4_lost_enrich, mixed_lost_enrich)
ggplot(combine_lost_enrich, aes(x=rowID, y= logFDR)) +
  geom_col(fill="#e67050", alpha=.6) +
  geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
  geom_vline(xintercept = c(-nrow(fluke_lost_enrich)-1, -nrow(fluke_lost_enrich)-nrow(C4_lost_enrich)-2)) + 
  scale_x_continuous(breaks=combine_lost_enrich$rowID, labels=combine_lost_enrich$ID) +
  theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))



####### plot combined gsea results
# gain
if (nrow(fluke_gain_gsea)>0) {
  fluke_gain_gsea$rowID = -1 : - nrow(fluke_gain_gsea)
}
if (nrow(C4_gain_gsea)>0) {
  C4_gain_gsea$rowID = -1 : - nrow(C4_gain_gsea)
  C4_gain_gsea$rowID = C4_gain_gsea$rowID - nrow(fluke_gain_gsea) - 1
}
if (nrow(mixed_gain_gsea)>0) {
  mixed_gain_gsea$rowID = -1 : - nrow(mixed_gain_gsea)
  mixed_gain_gsea$rowID = mixed_gain_gsea$rowID - nrow(fluke_gain_gsea) - 1 - nrow(C4_gain_gsea) - 1 
}
combine_gain_gsea = rbind(fluke_gain_gsea, C4_gain_gsea, mixed_gain_gsea)
ggplot(combine_gain_gsea, aes(x=rowID, y= logFDR)) +
  geom_col(fill="#e67050", alpha=.6) +
  geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
  geom_vline(xintercept = c(-nrow(fluke_gain_gsea)-1, -nrow(fluke_gain_gsea)-nrow(C4_gain_gsea)-2)) + 
  scale_x_continuous(breaks=combine_gain_gsea$rowID, labels=combine_gain_gsea$ID) +
  theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))


# lost
if (nrow(fluke_lost_gsea)>0) {
  fluke_lost_gsea$rowID = -1 : - nrow(fluke_lost_gsea)
}
if (nrow(C4_lost_gsea)>0) {
  C4_lost_gsea$rowID = -1 : - nrow(C4_lost_gsea)
  C4_lost_gsea$rowID = C4_lost_gsea$rowID - nrow(fluke_lost_gsea) - 1
}
if (nrow(mixed_lost_gsea)>0) {
  mixed_lost_gsea$rowID = -1 : - nrow(mixed_lost_gsea)
  mixed_lost_gsea$rowID = mixed_lost_gsea$rowID - nrow(fluke_lost_gsea) - 1 - nrow(C4_lost_gsea) - 1 
}
combine_lost_gsea = rbind(fluke_lost_gsea, C4_lost_gsea, mixed_lost_gsea)
ggplot(combine_lost_gsea, aes(x=rowID, y= logFDR)) +
  geom_col(fill="#e67050", alpha=.6) +
  geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
  geom_vline(xintercept = c(-nrow(fluke_lost_gsea)-1, -nrow(fluke_lost_gsea)-nrow(C4_lost_gsea)-2)) + 
  scale_x_continuous(breaks=combine_lost_gsea$rowID, labels=combine_lost_gsea$ID) +
  theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))



####### plot combined enrich (fluke, mixed) & gsea (c4) results
# gain
if (nrow(fluke_gain_enrich)>0) {
  fluke_gain_enrich$rowID = -1 : - nrow(fluke_gain_enrich)
}
if (nrow(C4_gain_gsea)>0) {
  C4_gain_gsea$rowID = -1 : - nrow(C4_gain_gsea)
  C4_gain_gsea$rowID = C4_gain_gsea$rowID - nrow(fluke_gain_enrich) - 1
}
if (nrow(mixed_gain_enrich)>0) {
  mixed_gain_enrich$rowID = -1 : - nrow(mixed_gain_enrich)
  mixed_gain_enrich$rowID = mixed_gain_enrich$rowID - nrow(fluke_gain_enrich) - 1 - nrow(C4_gain_gsea) - 1 
}
combine_gain_gsea = rbind(fluke_gain_enrich, C4_gain_gsea, mixed_gain_enrich)
ggplot(combine_gain_gsea, aes(x=rowID, y= logFDR)) +
  geom_col(fill="#e67050", alpha=.6) +
  geom_hline(yintercept = -log10(.05), linetype="dashed", color="grey30") +
  geom_vline(xintercept = c(-nrow(fluke_gain_enrich)-1, -nrow(fluke_gain_enrich)-nrow(C4_gain_gsea)-2)) + 
  scale_x_continuous(breaks=combine_gain_gsea$rowID, labels=combine_gain_gsea$ID) +
  theme_classic(base_size=14) + xlab("")  + ylab("-log10 FDR") + coord_flip() + theme(panel.border = element_rect(fill = NA))

```


```{r}
fluke_gain_enrich
fluke_lost_enrich
C4_gain_enrich
C4_lost_enrich
mixed_gain_enrich
mixed_lost_enrich
mixedClus3_gain_enrich
mixedClus3_lost_enrich
mixedAA_gain_enrich
mixedAA_lost_enrich
```

```{r}
fluke_gain_gsea
fluke_lost_gsea
C4_gain_gsea
C4_lost_gsea
mixed_gain_gsea
mixed_lost_gsea
mixedClus3_gain_gsea
mixedClus3_lost_gsea
mixedAA_gain_gsea
mixedAA_lost_gsea
```



### plot piechart of numbers of enhancers for each group
```{r}
########## get numbers of enhancers in each category
### gain
fluke_gain_all = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain$enhancer)
fluke_gain_all = fluke_gain_all[!is.na(fluke_gain_all)]

fluke_gain_val1 = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp$enhancer)
fluke_gain_val1 = fluke_gain_val1[!is.na(fluke_gain_val1)]

fluke_gain_val2 = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain$enhancer)
fluke_gain_val2 = fluke_gain_val2[!is.na(fluke_gain_val2)]

C4_gain_all = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain$enhancer)
C4_gain_all = C4_gain_all[!is.na(C4_gain_all)]

C4_gain_val1 = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp$enhancer)
C4_gain_val1 = C4_gain_val1[!is.na(C4_gain_val1)]

C4_gain_val2 = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain$enhancer)
C4_gain_val2 = C4_gain_val2[!is.na(C4_gain_val2)]

mixed_gain_all = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain$enhancer)
mixed_gain_all = mixed_gain_all[!is.na(mixed_gain_all)]

mixed_gain_val1 = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp$enhancer)
mixed_gain_val1 = mixed_gain_val1[!is.na(mixed_gain_val1)]

mixed_gain_val2 = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain$enhancer)
mixed_gain_val2 = mixed_gain_val2[!is.na(mixed_gain_val2)]

### lost
fluke_lost_all = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$lost$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$lost$enhancer)
fluke_lost_all = fluke_lost_all[!is.na(fluke_lost_all)]

fluke_lost_val1 = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$lost_exp$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$lost_exp$enhancer)
fluke_lost_val1 = fluke_lost_val1[!is.na(fluke_lost_val1)]

fluke_lost_val2 = intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$lost_exp_enhlost$enhancer, somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$lost_exp_enhlost$enhancer)
fluke_lost_val2 = fluke_lost_val2[!is.na(fluke_lost_val2)]

C4_lost_all = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$lost$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$lost$enhancer)
C4_lost_all = C4_lost_all[!is.na(C4_lost_all)]

C4_lost_val1 = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$lost_exp$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$lost_exp$enhancer)
C4_lost_val1 = C4_lost_val1[!is.na(C4_lost_val1)]

C4_lost_val2 = intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$lost_exp_enhlost$enhancer, somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$lost_exp_enhlost$enhancer)
C4_lost_val2 = C4_lost_val2[!is.na(C4_lost_val2)]

mixed_lost_all = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$lost$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$lost$enhancer)
mixed_lost_all = mixed_lost_all[!is.na(mixed_lost_all)]

mixed_lost_val1 = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$lost_exp$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$lost_exp$enhancer)
mixed_lost_val1 = mixed_lost_val1[!is.na(mixed_lost_val1)]

mixed_lost_val2 = intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$lost_exp_enhlost$enhancer, somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$lost_exp_enhlost$enhancer)
mixed_lost_val2 = mixed_lost_val2[!is.na(mixed_lost_val2)]

##### check each level of validation is a subset of the previous
# print("check each level of validation is a subset of the previous (all should be 0):")
# print(sum(!fluke_gain_val2 %in% fluke_gain_val1))
# print(sum(!fluke_gain_val1 %in% fluke_gain_all))
# print(sum(!C4_gain_val2 %in% C4_gain_val1))
# print(sum(!C4_gain_val1 %in% C4_gain_all))
# print(sum(!mixed_gain_val2 %in% mixed_gain_val1))
# print(sum(!mixed_gain_val1 %in% mixed_gain_all))
# print(sum(!fluke_lost_val2 %in% fluke_lost_val1))
# print(sum(!fluke_lost_val1 %in% fluke_lost_all))
# print(sum(!C4_lost_val2 %in% C4_lost_val1))
# print(sum(!C4_lost_val1 %in% C4_lost_all))
# print(sum(!mixed_lost_val2 %in% mixed_lost_val1))
# print(sum(!mixed_lost_val1 %in% mixed_lost_all))

##### check no overlap between groups
# print("check no overlap between groups (all should be 0):")
# print(length(intersect(fluke_gain_all, C4_gain_all))0)
# print(length(intersect(fluke_gain_all, mixed_gain_all)))
# print(length(intersect(mixed_gain_all, C4_gain_all)))
# print(length(intersect(fluke_lost_all, C4_lost_all)))
# print(length(intersect(fluke_lost_all, mixed_lost_all)))
# print(length(intersect(mixed_lost_all, C4_lost_all)))


gain_df = data.frame(name=rep(c("Unval", "Expr Val", "Expr & cellline val"), 3), group=c(rep("Fluke",3), rep("C4",3), rep("Mixed",3)), 
                     count=c(length(fluke_gain_all)-length(fluke_gain_val1), length(fluke_gain_val1)-length(fluke_gain_val2), length(fluke_gain_val2),
                             length(C4_gain_all)-length(C4_gain_val1), length(C4_gain_val1)-length(C4_gain_val2), length(C4_gain_val2),
                             length(mixed_gain_all)-length(mixed_gain_val1), length(mixed_gain_val1)-length(mixed_gain_val2), length(mixed_gain_val2))
                     )
print(gain_df)
print(paste0("total: ", sum(gain_df$count)))
print(paste0("fluke: ", sum(gain_df[gain_df$group=='Fluke', "count"]), ", ", sum(gain_df[gain_df$group=='Fluke', "count"])/sum(gain_df$count)*100, "%"))
print(paste0("C4: ", sum(gain_df[gain_df$group=='C4', "count"]), ", ", sum(gain_df[gain_df$group=='C4', "count"])/sum(gain_df$count)*100, "%"))
print(paste0("mixed: ", sum(gain_df[gain_df$group=='Mixed', "count"]), ", ", sum(gain_df[gain_df$group=='Mixed', "count"])/sum(gain_df$count)*100, "%"))

#### plot donut
#' x      numeric vector for each slice
#' group  vector identifying the group for each slice
#' labels vector of labels for individual slices
#' col    colors for each group
#' radius radius for inner and outer pie (usually in [0,1])
donuts <- function(x, group = 1, labels = NA, col = NULL, radius = c(.7, 1)) {
  group <- rep_len(group, length(x))
  ug  <- unique(group)
  tbl <- table(group)[order(ug)]

  col <- if (is.null(col))
    seq_along(ug) else rep_len(col, length(ug))
  col.main <- Map(rep, col[seq_along(tbl)], tbl)
  col.sub  <- lapply(col.main, function(x) {
    al <- head(seq(0, 1, length.out = length(x) + 2L)[-1L], -1L)
    Vectorize(adjustcolor)(x, alpha.f = al)
  })

  plot.new()

  par(new = TRUE)
  pie(x, border = NA, radius = radius[2L],
      col = unlist(col.sub), labels = labels)

  par(new = TRUE)
  pie(x, border = NA, radius = radius[1L],
      col = unlist(col.main), labels = NA)
}


# par(mfrow = c(1,2), mar = c(0,4,0,4))
donuts(gain_df$count, gain_df$group, col=c('blue', 'red2', 'green3'))

```



## Print enhancer locations bedfiles for Homer
```{r}
# Tumor_C1Flukepos_vs_Tumor_C2
zz = somatic_enhancers_groupspec$Tumor_C1Flukepos_vs_Tumor_C2$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C1Flukepos_vs_Tumor_C2.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")

# Tumor_C1Flukepos_vs_Tumor_C3
zz = somatic_enhancers_groupspec$Tumor_C1Flukepos_vs_Tumor_C3$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C1Flukepos_vs_Tumor_C3.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")


# Tumor_C1Flukepos_vs_Tumor_C4Flukeneg
zz = somatic_enhancers_groupspec$Tumor_C1Flukepos_vs_Tumor_C4Flukeneg$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C1Flukepos_vs_Tumor_C4Flukeneg.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")



# Tumor_C2_vs_Tumor_C1
zz = somatic_enhancers_groupspec$Tumor_C2_vs_Tumor_C1$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C2_vs_Tumor_C1.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")

# Tumor_C2_vs_Tumor_C3
zz = somatic_enhancers_groupspec$Tumor_C2_vs_Tumor_C3$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C2_vs_Tumor_C3.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")

# Tumor_C2_vs_Tumor_C4
zz = somatic_enhancers_groupspec$Tumor_C2_vs_Tumor_C4$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C2_vs_Tumor_C4.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")




# Tumor_C3_vs_Tumor_C1
zz = somatic_enhancers_groupspec$Tumor_C3_vs_Tumor_C1$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C3_vs_Tumor_C1.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")

# Tumor_C3_vs_Tumor_C2
zz = somatic_enhancers_groupspec$Tumor_C3_vs_Tumor_C2$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C3_vs_Tumor_C2.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")

# Tumor_C3_vs_Tumor_C4
zz = somatic_enhancers_groupspec$Tumor_C3_vs_Tumor_C4$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C3_vs_Tumor_C4.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")



# Tumor_C4FlukeNeg_vs_Tumor_C1FlukePos
zz = somatic_enhancers_groupspec$Tumor_C4FlukeNeg_vs_Tumor_C1FlukePos$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C4FlukeNeg_vs_Tumor_C1FlukePos.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")

# Tumor_C4FlukeNeg_vs_Tumor_C2
zz = somatic_enhancers_groupspec$Tumor_C4FlukeNeg_vs_Tumor_C2$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C4FlukeNeg_vs_Tumor_C2.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")


# Tumor_C4FlukeNeg_vs_Tumor_C3
zz = somatic_enhancers_groupspec$Tumor_C4FlukeNeg_vs_Tumor_C3$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C4FlukeNeg_vs_Tumor_C3.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")



# Tumor_AA_vs_Tumor_nonAA
zz = somatic_enhancers_groupspec$Tumor_AA_vs_Tumor_nonAA$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_AA_vs_Tumor_nonAA.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")


# Tumor_Flukepos_vs_Tumor_Flukeneg
zz = somatic_enhancers_groupspec$Tumor_Flukepos_vs_Tumor_Flukeneg$gain_exp_enhgain
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_Flukepos_vs_Tumor_Flukeneg.gain_exp_enhgain.bed", row.names = F, col.names=F, quote = F, sep="\t")



# Tumor_C1Flukepos_vs_Tumor_C2: gained estrogen response early genes
zz = somatic_enhancers_groupspec$Tumor_C1Flukepos_vs_Tumor_C2$gain_exp_enhgain
zz = zz[zz$gene %in% msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_EARLY, ]
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C1Flukepos_vs_Tumor_C2.gain_exp_enhgain.estrogen_early.bed", row.names = F, col.names=F, quote = F, sep="\t")

# estrogen late
zz = somatic_enhancers_groupspec$Tumor_C1Flukepos_vs_Tumor_C2$gain_exp_enhgain
zz = zz[zz$gene %in% msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_LATE, ]
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C1Flukepos_vs_Tumor_C2.gain_exp_enhgain.estrogen_late.bed", row.names = F, col.names=F, quote = F, sep="\t")


# Tumor_C1Flukepos_vs_Tumor_C3: gained estrogen response early genes
zz = somatic_enhancers_groupspec$Tumor_C1Flukepos_vs_Tumor_C3$gain_exp_enhgain
zz = zz[zz$gene %in% msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_EARLY, ]
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C1Flukepos_vs_Tumor_C3.gain_exp_enhgain.estrogen_early.bed", row.names = F, col.names=F, quote = F, sep="\t")

# estrogen late
zz = somatic_enhancers_groupspec$Tumor_C1Flukepos_vs_Tumor_C3$gain_exp_enhgain
zz = zz[zz$gene %in% msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_LATE, ]
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C1Flukepos_vs_Tumor_C3.gain_exp_enhgain.estrogen_late.bed", row.names = F, col.names=F, quote = F, sep="\t")


# Tumor_C1Flukepos_vs_Tumor_C4Flukeneg: gained estrogen response early genes
zz = somatic_enhancers_groupspec$Tumor_C1Flukepos_vs_Tumor_C4Flukeneg$gain_exp_enhgain
zz = zz[zz$gene %in% msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_EARLY, ]
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C1Flukepos_vs_Tumor_C4Flukeneg.gain_exp_enhgain.estrogen_early.bed", row.names = F, col.names=F, quote = F, sep="\t")

# estrogen late
zz = somatic_enhancers_groupspec$Tumor_C1Flukepos_vs_Tumor_C4Flukeneg$gain_exp_enhgain
zz = zz[zz$gene %in% msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_LATE, ]
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C1Flukepos_vs_Tumor_C4Flukeneg.gain_exp_enhgain.estrogen_late.bed", row.names = F, col.names=F, quote = F, sep="\t")


# Tumor_C4FlukeNeg_vs_Tumor_C1FlukePos: gained estrogen response early genes
zz = somatic_enhancers_groupspec$Tumor_C4FlukeNeg_vs_Tumor_C1FlukePos$gain_exp_enhgain
zz = zz[zz$gene %in% msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_EARLY, ]
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C4FlukeNeg_vs_Tumor_C1FlukePos.gain_exp_enhgain.estrogen_early.bed", row.names = F, col.names=F, quote = F, sep="\t")

# estrogen late
zz = somatic_enhancers_groupspec$Tumor_C4FlukeNeg_vs_Tumor_C1FlukePos$gain_exp_enhgain
zz = zz[zz$gene %in% msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_LATE, ]
zz = sapply(unique(zz$enhancer), function(ee){
  paste0(ee, "_", paste0(unique(zz[zz$enhancer==ee, "gene"]), collapse=","))
})
yy = data.frame(chr=sapply(strsplit(zz, split=":"), function(x){x[1]}), 
            start=sapply(strsplit(zz, split="[:_-]"), function(x){x[2]}), 
            end=sapply(strsplit(zz, split="[:_-]"), function(x){x[3]}), 
            id=zz, notused="1", strand="+", stringsAsFactors = F)
write.table(yy, file="C:/Users/cherny/Google Drive/CCA-epigenetics/CCA_enhancers_project/outputs/homer_bed.Tumor_C4FlukeNeg_vs_Tumor_C1FlukePos.gain_exp_enhgain.estrogen_late.bed", row.names = F, col.names=F, quote = F, sep="\t")
```




# print gained & lost enhancers and the significant pathways, for supplementary table
### Fluke
```{r}

### Fluke-specific gain
# get overlapping enhancer-gene pairs 
zz1 = somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain
zz1$enhancergene = paste0(zz1$enhancer, '|', zz1$gene)

zz2 = somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain
zz2$enhancergene = paste0(zz2$enhancer, '|', zz2$gene)

intersect_enhgene = intersect(zz1$enhancergene, zz2$enhancergene)

# keep only overlapping enhancer-genes 
zz1 = zz1[, c("gene", "enhancer", "enhancergene", "ConcPLW_1.x.1", "ConcPLW_2.x.1", "Fold.x.1", "p.value.x.1", "FDR.x.1", "ConcPLW_1.x.2", "ConcPLW_2.x.2", "Fold.x.2", "p.value.x.2", "FDR.x.2.shrunk", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2", "ConcPLW_1.y.1", "ConcPLW_2.y.1", "Fold.y.1")]
zz1 = zz1[zz1$enhancergene %in% intersect_enhgene,]

zz2 = zz2[, c("gene", "enhancer", "enhancergene", "ConcPLW_1.x.1", "ConcPLW_2.x.1", "Fold.x.1", "p.value.x.1", "FDR.x.1", "ConcPLW_1.x.2", "ConcPLW_2.x.2", "Fold.x.2", "p.value.x.2", "FDR.x.2.shrunk", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2", "ConcPLW_1.y.1", "ConcPLW_2.y.1", "Fold.y.1")]
zz2 = zz2[zz2$enhancergene %in% intersect_enhgene,]

# merge
zz = merge(zz1, zz2, by="enhancergene")
zz = zz[, c("enhancer.x", "gene.x", 
            "Fold.x.1.x", "p.value.x.1.x", "FDR.x.1.x", 
            "Fold.x.2.x", "p.value.x.2.x", "FDR.x.2.shrunk.x", 
            "Fold.x.2.y", "p.value.x.2.y", "FDR.x.2.shrunk.y", 
            "expr_fold.x.1.x", "expr_p.x.1.x", 
            "expr_fold.x.2.x", "expr_p.x.2.x", 
            "expr_fold.x.2.y", "expr_p.x.2.y",
            "Fold.y.1.x")]
colnames(zz) = c("enhancer", "gene", 
                 "enh Fluke vs Norm (logFC)", "enh Fluke vs Norm (P)", "enh Fluke vs Norm (FDR)", 
                 "enh Fluke vs C4 (logFC)", "enh Fluke vs C4 (P)", "enh Fluke vs C4 (FDR)", 
                 "enh Fluke vs Mixed (logFC)", "enh Fluke vs Mixed (P)", "enh Fluke vs Mixed (FDR)", 
                 "exp Fluke vs Norm (logFC)", "exp Fluke vs Norm (P)", 
                 "exp Fluke vs C4 (logFC)", "exp Fluke vs C4 (P)", 
                 "exp Fluke vs Mixed (logFC)", "exp Fluke vs Mixed (P)",
                 "enh cell Fluke vs Norm (logFC)")

zz$HALLMARK_ESTROGEN_RESPONSE_EARLY = zz$gene %in% msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_EARLY
zz$HALLMARK_ESTROGEN_RESPONSE_LATE = zz$gene %in% msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_LATE
zz$HALLMARK_APICAL_JUNCTION = zz$gene %in% msigdb_H_list$HALLMARK_APICAL_JUNCTION
zz$HALLMARK_GLYCOLYSIS = zz$gene %in% msigdb_H_list$HALLMARK_GLYCOLYSIS
zz$HALLMARK_KRAS_SIGNALING_UP = zz$gene %in% msigdb_H_list$HALLMARK_KRAS_SIGNALING_UP
write.table(zz, file="somatic_enhancers_groupspec_Fluke_gain.txt", row.names = F, col.names=T, quote = F, sep="\t")



### Fluke-specific lost
# get overlapping enhancer-gene pairs 
zz1 = somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$lost_exp_enhlost
zz1$enhancergene = paste0(zz1$enhancer, '|', zz1$gene)

zz2 = somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$lost_exp_enhlost
zz2$enhancergene = paste0(zz2$enhancer, '|', zz2$gene)

intersect_enhgene = intersect(zz1$enhancergene, zz2$enhancergene)

# keep only overlapping enhancer-genes 
zz1 = zz1[, c("gene", "enhancer", "enhancergene", "ConcPLW_1.x.1", "ConcPLW_2.x.1", "Fold.x.1", "p.value.x.1", "FDR.x.1", "ConcPLW_1.x.2", "ConcPLW_2.x.2", "Fold.x.2", "p.value.x.2", "FDR.x.2.shrunk", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2", "ConcPLW_1.y.1", "ConcPLW_2.y.1", "Fold.y.1")]
zz1 = zz1[zz1$enhancergene %in% intersect_enhgene,]

zz2 = zz2[, c("gene", "enhancer", "enhancergene", "ConcPLW_1.x.1", "ConcPLW_2.x.1", "Fold.x.1", "p.value.x.1", "FDR.x.1", "ConcPLW_1.x.2", "ConcPLW_2.x.2", "Fold.x.2", "p.value.x.2", "FDR.x.2.shrunk", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2", "ConcPLW_1.y.1", "ConcPLW_2.y.1", "Fold.y.1")]
zz2 = zz2[zz2$enhancergene %in% intersect_enhgene,]

# merge
zz = merge(zz1, zz2, by="enhancergene")
zz = zz[, c("enhancer.x", "gene.x", 
            "Fold.x.1.x", "p.value.x.1.x", "FDR.x.1.x", 
            "Fold.x.2.x", "p.value.x.2.x", "FDR.x.2.shrunk.x", 
            "Fold.x.2.y", "p.value.x.2.y", "FDR.x.2.shrunk.y", 
            "expr_fold.x.1.x", "expr_p.x.1.x", 
            "expr_fold.x.2.x", "expr_p.x.2.x", 
            "expr_fold.x.2.y", "expr_p.x.2.y",
            "Fold.y.1.x")]
colnames(zz) = c("enhancer", "gene", 
                 "enh Fluke vs Norm (logFC)", "enh Fluke vs Norm (P)", "enh Fluke vs Norm (FDR)", 
                 "enh Fluke vs C4 (logFC)", "enh Fluke vs C4 (P)", "enh Fluke vs C4 (FDR)", 
                 "enh Fluke vs Mixed (logFC)", "enh Fluke vs Mixed (P)", "enh Fluke vs Mixed (FDR)", 
                 "exp Fluke vs Norm (logFC)", "exp Fluke vs Norm (P)", 
                 "exp Fluke vs C4 (logFC)", "exp Fluke vs C4 (P)", 
                 "exp Fluke vs Mixed (logFC)", "exp Fluke vs Mixed (P)",
                 "enh cell Fluke vs Norm (logFC)")

zz$HALLMARK_UV_RESPONSE_DN = zz$gene %in% msigdb_H_list$HALLMARK_UV_RESPONSE_DN
write.table(zz, file="somatic_enhancers_groupspec_Fluke_lost.txt", row.names = F, col.names=T, quote = F, sep="\t")

```



### c4
```{r}

### c4-specific gain
# get overlapping enhancer-gene pairs 
zz1 = somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain
zz1$enhancergene = paste0(zz1$enhancer, '|', zz1$gene)

zz2 = somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain
zz2$enhancergene = paste0(zz2$enhancer, '|', zz2$gene)

intersect_enhgene = intersect(zz1$enhancergene, zz2$enhancergene)

# keep only overlapping enhancer-genes 
zz1 = zz1[, c("gene", "enhancer", "enhancergene", "ConcPLW_1.x.1", "ConcPLW_2.x.1", "Fold.x.1", "p.value.x.1", "FDR.x.1", "ConcPLW_1.x.2", "ConcPLW_2.x.2", "Fold.x.2", "p.value.x.2", "FDR.x.2.shrunk", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2", "ConcPLW_1.y.1", "ConcPLW_2.y.1", "Fold.y.1")]
zz1 = zz1[zz1$enhancergene %in% intersect_enhgene,]

zz2 = zz2[, c("gene", "enhancer", "enhancergene", "ConcPLW_1.x.1", "ConcPLW_2.x.1", "Fold.x.1", "p.value.x.1", "FDR.x.1", "ConcPLW_1.x.2", "ConcPLW_2.x.2", "Fold.x.2", "p.value.x.2", "FDR.x.2.shrunk", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2", "ConcPLW_1.y.1", "ConcPLW_2.y.1", "Fold.y.1")]
zz2 = zz2[zz2$enhancergene %in% intersect_enhgene,]

# merge
zz = merge(zz1, zz2, by="enhancergene")
zz = zz[, c("enhancer.x", "gene.x", 
            "Fold.x.1.x", "p.value.x.1.x", "FDR.x.1.x", 
            "Fold.x.2.x", "p.value.x.2.x", "FDR.x.2.shrunk.x", 
            "Fold.x.2.y", "p.value.x.2.y", "FDR.x.2.shrunk.y", 
            "expr_fold.x.1.x", "expr_p.x.1.x", 
            "expr_fold.x.2.x", "expr_p.x.2.x", 
            "expr_fold.x.2.y", "expr_p.x.2.y",
            "Fold.y.1.x")]
colnames(zz) = c("enhancer", "gene", 
                 "enh C4 vs Norm (logFC)", "enh C4 vs Norm (P)", "enh C4 vs Norm (FDR)", 
                 "enh C4 vs Fluke (logFC)", "enh C4 vs Fluke (P)", "enh C4 vs Fluke (FDR)", 
                 "enh C4 vs Mixed (logFC)", "enh C4 vs Mixed (P)", "enh C4 vs Mixed (FDR)", 
                 "exp C4 vs Norm (logFC)", "exp C4 vs Norm (P)", 
                 "exp C4 vs Fluke (logFC)", "exp C4 vs Fluke (P)", 
                 "exp C4 vs Mixed (logFC)", "exp C4 vs Mixed (P)",
                 "enh cell C4 vs Norm (logFC)")

zz$HALLMARK_OXIDATIVE_PHOSPHORYLATION = zz$gene %in% msigdb_H_list$HALLMARK_OXIDATIVE_PHOSPHORYLATION
write.table(zz, file="somatic_enhancers_groupspec_C4_gain.txt", row.names = F, col.names=T, quote = F, sep="\t")



### c4-specific lost
# get overlapping enhancer-gene pairs 
zz1 = somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$lost_exp_enhlost
zz1$enhancergene = paste0(zz1$enhancer, '|', zz1$gene)

zz2 = somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$lost_exp_enhlost
zz2$enhancergene = paste0(zz2$enhancer, '|', zz2$gene)

intersect_enhgene = intersect(zz1$enhancergene, zz2$enhancergene)

# keep only overlapping enhancer-genes 
zz1 = zz1[, c("gene", "enhancer", "enhancergene", "ConcPLW_1.x.1", "ConcPLW_2.x.1", "Fold.x.1", "p.value.x.1", "FDR.x.1", "ConcPLW_1.x.2", "ConcPLW_2.x.2", "Fold.x.2", "p.value.x.2", "FDR.x.2.shrunk", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2", "ConcPLW_1.y.1", "ConcPLW_2.y.1", "Fold.y.1")]
zz1 = zz1[zz1$enhancergene %in% intersect_enhgene,]

zz2 = zz2[, c("gene", "enhancer", "enhancergene", "ConcPLW_1.x.1", "ConcPLW_2.x.1", "Fold.x.1", "p.value.x.1", "FDR.x.1", "ConcPLW_1.x.2", "ConcPLW_2.x.2", "Fold.x.2", "p.value.x.2", "FDR.x.2.shrunk", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2", "ConcPLW_1.y.1", "ConcPLW_2.y.1", "Fold.y.1")]
zz2 = zz2[zz2$enhancergene %in% intersect_enhgene,]

# merge
zz = merge(zz1, zz2, by="enhancergene")
zz = zz[, c("enhancer.x", "gene.x", 
            "Fold.x.1.x", "p.value.x.1.x", "FDR.x.1.x", 
            "Fold.x.2.x", "p.value.x.2.x", "FDR.x.2.shrunk.x", 
            "Fold.x.2.y", "p.value.x.2.y", "FDR.x.2.shrunk.y", 
            "expr_fold.x.1.x", "expr_p.x.1.x", 
            "expr_fold.x.2.x", "expr_p.x.2.x", 
            "expr_fold.x.2.y", "expr_p.x.2.y",
            "Fold.y.1.x")]
colnames(zz) = c("enhancer", "gene", 
                 "enh C4 vs Norm (logFC)", "enh C4 vs Norm (P)", "enh C4 vs Norm (FDR)", 
                 "enh C4 vs Fluke (logFC)", "enh C4 vs Fluke (P)", "enh C4 vs Fluke (FDR)", 
                 "enh C4 vs Mixed (logFC)", "enh C4 vs Mixed (P)", "enh C4 vs Mixed (FDR)", 
                 "exp C4 vs Norm (logFC)", "exp C4 vs Norm (P)", 
                 "exp C4 vs Fluke (logFC)", "exp C4 vs Fluke (P)", 
                 "exp C4 vs Mixed (logFC)", "exp C4 vs Mixed (P)",
                 "enh cell C4 vs Norm (logFC)")

zz$HALLMARK_P53_PATHWAY = zz$gene %in% msigdb_H_list$HALLMARK_P53_PATHWAY
zz$HALLMARK_ESTROGEN_RESPONSE_LATE = zz$gene %in% msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_LATE
zz$HALLMARK_ESTROGEN_RESPONSE_EARLY = zz$gene %in% msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_EARLY
zz$HALLMARK_ANDROGEN_RESPONSE = zz$gene %in% msigdb_H_list$HALLMARK_ANDROGEN_RESPONSE
zz$HALLMARK_SPERMATOGENESIS = zz$gene %in% msigdb_H_list$HALLMARK_SPERMATOGENESIS
zz$HALLMARK_COMPLEMENT = zz$gene %in% msigdb_H_list$HALLMARK_COMPLEMENT
zz$HALLMARK_IL2_STAT5_SIGNALING = zz$gene %in% msigdb_H_list$HALLMARK_IL2_STAT5_SIGNALING
write.table(zz, file="somatic_enhancers_groupspec_C4_lost.txt", row.names = F, col.names=T, quote = F, sep="\t")



```


### Mixed
```{r}

### Mixed-specific gain
# get overlapping enhancer-gene pairs 
zz1 = somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain
zz1$enhancergene = paste0(zz1$enhancer, '|', zz1$gene)

zz2 = somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain
zz2$enhancergene = paste0(zz2$enhancer, '|', zz2$gene)

intersect_enhgene = intersect(zz1$enhancergene, zz2$enhancergene)

# keep only overlapping enhancer-genes 
zz1 = zz1[, c("gene", "enhancer", "enhancergene", "ConcPLW_1.x.1", "ConcPLW_2.x.1", "Fold.x.1", "p.value.x.1", "FDR.x.1", "ConcPLW_1.x.2", "ConcPLW_2.x.2", "Fold.x.2", "p.value.x.2", "FDR.x.2.shrunk", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2", "ConcPLW_1.y.1", "ConcPLW_2.y.1", "Fold.y.1")]
zz1 = zz1[zz1$enhancergene %in% intersect_enhgene,]

zz2 = zz2[, c("gene", "enhancer", "enhancergene", "ConcPLW_1.x.1", "ConcPLW_2.x.1", "Fold.x.1", "p.value.x.1", "FDR.x.1", "ConcPLW_1.x.2", "ConcPLW_2.x.2", "Fold.x.2", "p.value.x.2", "FDR.x.2.shrunk", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2", "ConcPLW_1.y.1", "ConcPLW_2.y.1", "Fold.y.1")]
zz2 = zz2[zz2$enhancergene %in% intersect_enhgene,]

# merge
zz = merge(zz1, zz2, by="enhancergene")
zz = zz[, c("enhancer.x", "gene.x", 
            "Fold.x.1.x", "p.value.x.1.x", "FDR.x.1.x", 
            "Fold.x.2.x", "p.value.x.2.x", "FDR.x.2.shrunk.x", 
            "Fold.x.2.y", "p.value.x.2.y", "FDR.x.2.shrunk.y", 
            "expr_fold.x.1.x", "expr_p.x.1.x", 
            "expr_fold.x.2.x", "expr_p.x.2.x", 
            "expr_fold.x.2.y", "expr_p.x.2.y",
            "Fold.y.1.x")]
colnames(zz) = c("enhancer", "gene", 
                 "enh Mixed vs Norm (logFC)", "enh Mixed vs Norm (P)", "enh Mixed vs Norm (FDR)", 
                 "enh Mixed vs Fluke (logFC)", "enh Mixed vs Fluke (P)", "enh Mixed vs Fluke (FDR)", 
                 "enh Mixed vs C4 (logFC)", "enh Mixed vs C4 (P)", "enh Mixed vs C4 (FDR)", 
                 "exp Mixed vs Norm (logFC)", "exp Mixed vs Norm (P)", 
                 "exp Mixed vs Fluke (logFC)", "exp Mixed vs Fluke (P)", 
                 "exp Mixed vs C4 (logFC)", "exp Mixed vs C4 (P)",
                 "enh cell Mixed vs Norm (logFC)")

zz$HALLMARK_INTERFERON_GAMMA_RESPONSE = zz$gene %in% msigdb_H_list$HALLMARK_INTERFERON_GAMMA_RESPONSE
zz$HALLMARK_ALLOGRAFT_REJECTION = zz$gene %in% msigdb_H_list$HALLMARK_ALLOGRAFT_REJECTION
zz$HALLMARK_INTERFERON_ALPHA_RESPONSE = zz$gene %in% msigdb_H_list$HALLMARK_INTERFERON_ALPHA_RESPONSE
zz$HALLMARK_COAGULATION = zz$gene %in% msigdb_H_list$HALLMARK_COAGULATION
zz$HALLMARK_COMPLEMENT = zz$gene %in% msigdb_H_list$HALLMARK_COMPLEMENT
zz$HALLMARK_XENOBIOTIC_METABOLISM = zz$gene %in% msigdb_H_list$HALLMARK_XENOBIOTIC_METABOLISM
zz$HALLMARK_IL2_STAT5_SIGNALING = zz$gene %in% msigdb_H_list$HALLMARK_IL2_STAT5_SIGNALING
zz$HALLMARK_IL6_JAK_STAT3_SIGNALING = zz$gene %in% msigdb_H_list$HALLMARK_IL6_JAK_STAT3_SIGNALING
zz$HALLMARK_INFLAMMATORY_RESPONSE = zz$gene %in% msigdb_H_list$HALLMARK_INFLAMMATORY_RESPONSE
write.table(zz, file="somatic_enhancers_groupspec_Mixed_gain.txt", row.names = F, col.names=T, quote = F, sep="\t")



### Mixed-specific lost
# get overlapping enhancer-gene pairs 
zz1 = somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$lost_exp_enhlost
zz1$enhancergene = paste0(zz1$enhancer, '|', zz1$gene)

zz2 = somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$lost_exp_enhlost
zz2$enhancergene = paste0(zz2$enhancer, '|', zz2$gene)

intersect_enhgene = intersect(zz1$enhancergene, zz2$enhancergene)

# keep only overlapping enhancer-genes 
zz1 = zz1[, c("gene", "enhancer", "enhancergene", "ConcPLW_1.x.1", "ConcPLW_2.x.1", "Fold.x.1", "p.value.x.1", "FDR.x.1", "ConcPLW_1.x.2", "ConcPLW_2.x.2", "Fold.x.2", "p.value.x.2", "FDR.x.2.shrunk", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2", "ConcPLW_1.y.1", "ConcPLW_2.y.1", "Fold.y.1")]
zz1 = zz1[zz1$enhancergene %in% intersect_enhgene,]

zz2 = zz2[, c("gene", "enhancer", "enhancergene", "ConcPLW_1.x.1", "ConcPLW_2.x.1", "Fold.x.1", "p.value.x.1", "FDR.x.1", "ConcPLW_1.x.2", "ConcPLW_2.x.2", "Fold.x.2", "p.value.x.2", "FDR.x.2.shrunk", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2", "ConcPLW_1.y.1", "ConcPLW_2.y.1", "Fold.y.1")]
zz2 = zz2[zz2$enhancergene %in% intersect_enhgene,]

# merge
zz = merge(zz1, zz2, by="enhancergene")
zz = zz[, c("enhancer.x", "gene.x", 
            "Fold.x.1.x", "p.value.x.1.x", "FDR.x.1.x", 
            "Fold.x.2.x", "p.value.x.2.x", "FDR.x.2.shrunk.x", 
            "Fold.x.2.y", "p.value.x.2.y", "FDR.x.2.shrunk.y", 
            "expr_fold.x.1.x", "expr_p.x.1.x", 
            "expr_fold.x.2.x", "expr_p.x.2.x", 
            "expr_fold.x.2.y", "expr_p.x.2.y",
            "Fold.y.1.x")]
colnames(zz) = c("enhancer", "gene", 
                 "enh Mixed vs Norm (logFC)", "enh Mixed vs Norm (P)", "enh Mixed vs Norm (FDR)", 
                 "enh Mixed vs Fluke (logFC)", "enh Mixed vs Fluke (P)", "enh Mixed vs Fluke (FDR)", 
                 "enh Mixed vs C4 (logFC)", "enh Mixed vs C4 (P)", "enh Mixed vs C4 (FDR)", 
                 "exp Mixed vs Norm (logFC)", "exp Mixed vs Norm (P)", 
                 "exp Mixed vs Fluke (logFC)", "exp Mixed vs Fluke (P)", 
                 "exp Mixed vs C4 (logFC)", "exp Mixed vs C4 (P)",
                 "enh cell Mixed vs Norm (logFC)")

write.table(zz, file="somatic_enhancers_groupspec_Mixed_lost.txt", row.names = F, col.names=T, quote = F, sep="\t")





```
