---
title: "CCA tissue enhancer analysis: Cell lines"
output: 
  html_notebook:
    toc: true
    toc_float: true

---



  
# Tissue enhancers

  
  

```{r include=F}

# options(scipen=0)
rm(list=ls())
library(DiffBind, warn.conflicts=F, quietly=T)
library(beeswarm)
library(msigdbr)
library(fgsea)
library(gplots)
library(clusterProfiler)
library(biomaRt)
source('analysis_cca_enhancers_functions.R')


# read tissue enhancer data
load(file="Rdata_cca_k27ac_0_prelim_final.Rdata")
rm(K27ac_enhsup_dba, K27ac_origPeaks, superenhancers)


# read diffbind results
# load(file="Rdata_cca_k27ac_3_cells_diffbind.Rdata")
# rm(K27ac_cells_dba, K27ac_cells_enhsup_tisspeaks_dba)
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts1.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts2.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts3.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts4.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts5.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts1_diffexpr.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts2_diffexpr.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts3_diffexpr.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts4_diffexpr.Rdata")
# load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts5_diffexpr.Rdata")


#  load multicomp results
load("Rdata_cca_k27ac_3a_somatic_enhancers.Rdata")
load("Rdata_cca_k27ac_3b_somatic_enhancers_groupspec.Rdata")


# read mapping
target_map = read.table("target_gene_mapping/20190628_enhancer_target_map_noTAD_CPLW_rho.3_liberal.txt", header=T, stringsAsFactors = F) 
target_map = target_map[!is.na(target_map$gene),c("enhancer", "gene")]
target_map = target_map[!duplicated(target_map),]




# read mSigDB genesets
msigdb_H = msigdbr(species = "Homo sapiens", category = "H") # 50 genesets
msigdb_H_list = msigdb_H %>% split(x = .$gene_symbol, f = .$gs_name)
msigdb_H_term2gene = data.frame(TERM=msigdb_H$gs_name, GENE=msigdb_H$gene_symbol, stringsAsFactors = F)
msigdb_H_term2gene = rbind(msigdb_H_term2gene, data.frame(TERM="ALL_GENES", GENE=unique(na.omit(target_map$gene)), stringsAsFactors = F)) # clusterProfiler considers universe = genes in genesets


# read biomart gene descriptions
biomart_ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")


get_genesets = function(gene, pathways=c("H","R")) {
  ann = NULL
  if ("H" %in% pathways) {
    ann = c(ann, msigdb_H_term2gene[msigdb_H_term2gene$GENE==gene,"TERM"])
  }
  if ("R" %in% pathways) {
    ann = c(ann, msigdb_R_term2gene[msigdb_R_term2gene$GENE==gene,"TERM"])
  }
  ann = unique(ann)
  ann = ann[ann!="ALL_GENES"]
  ann = gsub("HALLMARK", "H", ann)
  ann = gsub("REACTOME", "R", ann)
  ann = sapply(ann, function(x){substr(x, 1, 30)})
  paste0(ann, collapse=",")
}
get_gene_descr = function(gene){
  descr = getBM(attributes=c('description'), filters='external_gene_name', values=gene, mart=biomart_ensembl)[1,1]
  if (is.na(descr)) { "" 
  } else {  strsplit(descr, split=" [", fixed=T)[[1]][1] }
}



samples_data = read.table("2021.08.22 samples_info.txt", stringsAsFactors = F, header=T, sep="\t", check.names = F)
```




### plot tissue enhancers


```{r fig.width=12}


####### get deseq2 normalization factors (exclude liver samples)
# # get the factors based on the diffbind-default-normalization matrix
# # note: Supposed to use raw counts! but somehow this gives nice heatmap?!
# deseq2facs = get_deseq2_normfacs(K27ac_dba, samps=K27ac_dba$samples[K27ac_dba$samples$Tissue != 'Normal_liver',"SampleID"])

# get factors based on raw counts, BUT SET NEGATIVES TO 0 FIRST!
zz_K27ac_dba = dba.count(K27ac_dba, peaks=NULL, score=DBA_SCORE_READS_MINUS)
zz_K27ac_data = zz_K27ac_dba$binding[, -1:-3]
zz_K27ac_data = zz_K27ac_data[, colnames(zz_K27ac_data) %in% samples_data[samples_data$EGroup!="Normal_liver","Sample"]]
zz_K27ac_data[zz_K27ac_data < 0] = 0
deseq2facs = estimateSizeFactorsForMatrix(zz_K27ac_data)
rm(zz_K27ac_dba, zz_K27ac_data)

geneset_genes = unique(c(msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_LATE, msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_EARLY))
# geneset_genes2 = msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_EARLY



####### get enhancers & genes gained in Fluke vs C4 & Mixed
zz_fluke_c4 = somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain[, c("gene", "enhancer", "Fold.x.1", "p.value.x.1", "Fold.x.2", "p.value.x.2", "Fold.y.1", "p.value.y.1", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2")]

zz_fluke_mix = somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain[, c("gene", "enhancer", "Fold.x.1", "p.value.x.1", "Fold.x.2", "p.value.x.2", "Fold.y.1", "p.value.y.1", "expr_fold.x.1", "expr_p.x.1", "expr_fold.x.2", "expr_p.x.2")]

fluke_gain = merge(zz_fluke_c4, zz_fluke_mix, by=c('enhancer','gene'), suffixes=c('.C4', '.Mixed'), sort=F)

# for each gene, select one enhancer
fluke_gain_split = split(fluke_gain, fluke_gain$gene)
set.seed(123)
zz = lapply(fluke_gain_split, function(x){return (x[sample(1:nrow(x), size=1), ])}) # select random enhancer 
fluke_gain = do.call(rbind, zz)
set.seed(777)
fluke_gain = fluke_gain[sample(nrow(fluke_gain)),] # randomize order because somatic_enhancers_groupspec puts genes with no cell-line expression at bottom
fluke_gain$enhancergene = paste0(fluke_gain$enhancer, '|', fluke_gain$gene)
fluke_enhancergene = fluke_gain[, c("enhancergene", "enhancer", "gene")]



######### get enhancers & genes NOT GAINED in fluke vs normal, fluke vs C4, fluke vs mixed (enhancer & expression)

### fluke vs normal
contrast = 1
# get diffbind results for all enhancer
my_report = data.frame(dba.report(K27ac_dba_deseq_contrasts1, contrast=contrast, method=DBA_DESEQ2, th=1, bCounts = T, bNormalized = T), check.names = F)
colnames(my_report)[6:ncol(my_report)] = colnames(mcols(dba.report(K27ac_dba_deseq_contrasts1, contrast=contrast, method=DBA_DESEQ2, th=0.05, bCounts = T, bNormalized = T)))
my_report = my_report[!is.na(my_report$`p-value`) & !is.na(my_report$Fold),]
my_report$enhancer = paste0(my_report$seqnames, ":", my_report$start, "-", my_report$end)
# merge with gene expression
my_report = merge(my_report, target_map, by="enhancer", sort=F)
diffexpr = diffexpr_list_contrasts1[[contrast]]$diffexpr
diffexpr = data.frame(gene=rownames(diffexpr), expr_fold=diffexpr$log2FoldChange, expr_p=diffexpr$pvalue, stringsAsFactors = F)
diffexpr = diffexpr[!is.na(diffexpr$expr_fold) & !is.na(diffexpr$expr_p),]
my_report = merge(my_report, diffexpr, by="gene", sort=F)
my_report$enhancergene = paste0(my_report$enhancer, "|", my_report$gene)
# get nogain ehhancers
fluke_norm_nogain = my_report[my_report$`p-value` > .3 & my_report$expr_p > .3,]

### fluke vs C4
contrast = 3
# get diffbind results for all enhancer
my_report = data.frame(dba.report(K27ac_dba_deseq_contrasts1, contrast=contrast, method=DBA_DESEQ2, th=1, bCounts = T, bNormalized = T), check.names = F)
colnames(my_report)[6:ncol(my_report)] = colnames(mcols(dba.report(K27ac_dba_deseq_contrasts1, contrast=contrast, method=DBA_DESEQ2, th=0.05, bCounts = T, bNormalized = T)))
my_report = my_report[!is.na(my_report$`p-value`) & !is.na(my_report$Fold),]
my_report$enhancer = paste0(my_report$seqnames, ":", my_report$start, "-", my_report$end)
# merge with gene expression
my_report = merge(my_report, target_map, by="enhancer", sort=F)
diffexpr = diffexpr_list_contrasts1[[contrast]]$diffexpr
diffexpr = data.frame(gene=rownames(diffexpr), expr_fold=diffexpr$log2FoldChange, expr_p=diffexpr$pvalue, stringsAsFactors = F)
diffexpr = diffexpr[!is.na(diffexpr$expr_fold) & !is.na(diffexpr$expr_p),]
my_report = merge(my_report, diffexpr, by="gene", sort=F)
my_report$enhancergene = paste0(my_report$enhancer, "|", my_report$gene)
# get nogain ehhancers
fluke_C4_nogain = my_report[my_report$`p-value` > .3 & my_report$expr_p > .3,]

### fluke vs mixed
contrast = 4
# get diffbind results for all enhancer
my_report = data.frame(dba.report(K27ac_dba_deseq_contrasts1, contrast=contrast, method=DBA_DESEQ2, th=1, bCounts = T, bNormalized = T), check.names = F)
colnames(my_report)[6:ncol(my_report)] = colnames(mcols(dba.report(K27ac_dba_deseq_contrasts1, contrast=contrast, method=DBA_DESEQ2, th=0.05, bCounts = T, bNormalized = T)))
my_report = my_report[!is.na(my_report$`p-value`) & !is.na(my_report$Fold),]
my_report$enhancer = paste0(my_report$seqnames, ":", my_report$start, "-", my_report$end)
# merge with gene expression
my_report = merge(my_report, target_map, by="enhancer", sort=F)
diffexpr = diffexpr_list_contrasts1[[contrast]]$diffexpr
diffexpr = data.frame(gene=rownames(diffexpr), expr_fold=diffexpr$log2FoldChange, expr_p=diffexpr$pvalue, stringsAsFactors = F)
diffexpr = diffexpr[!is.na(diffexpr$expr_fold) & !is.na(diffexpr$expr_p),]
my_report = merge(my_report, diffexpr, by="gene", sort=F)
my_report$enhancergene = paste0(my_report$enhancer, "|", my_report$gene)
# get nogain ehhancers
fluke_mixed_nogain = my_report[my_report$`p-value` > .3 & my_report$expr_p > .3,]

fluke_nogain_enhancergene = intersect(intersect(fluke_norm_nogain$enhancergene, fluke_C4_nogain$enhancergene), fluke_mixed_nogain$enhancergene)
fluke_nogain_enhancergene = data.frame(enhancergene=fluke_nogain_enhancergene, stringsAsFactors = F)
fluke_nogain_enhancergene = cbind(fluke_nogain_enhancergene, do.call(rbind, strsplit(fluke_nogain_enhancergene$enhancergene, '|', fixed=T)), stringsAsFactors=F)
colnames(fluke_nogain_enhancergene)[2:3] = c("enhancer", "gene")
fluke_nogain_enhancergene = fluke_nogain_enhancergene[!duplicated(fluke_nogain_enhancergene$enhancer),] # remove duplicate enhancers

# for each gene, select one enhancer
fluke_nogain_enhancergene_split = split(fluke_nogain_enhancergene, fluke_nogain_enhancergene$gene)
set.seed(123)
zz = lapply(fluke_nogain_enhancergene_split, function(x){return (x[sample(1:nrow(x), size=1), ])}) # select random enhancer 
fluke_nogain_enhancergene = do.call(rbind, zz)
# select 200 to plot
set.seed(999)
fluke_nogain_enhancergene = fluke_nogain_enhancergene[sample(1:nrow(fluke_nogain_enhancergene), size=200),]


###### order the data to plot
plot_samples = samples_data[samples_data$EGroup!="Normal_liver",]
plot_samples = plot_samples[order(factor(plot_samples$EGroup, levels=c("Fluke", "C4", "Mixed", "Normal"))),]
plot_enhancers = rbind(fluke_enhancergene, fluke_nogain_enhancergene)
plot_enhancers$gain = F
plot_enhancers[plot_enhancers$enhancergene %in% fluke_enhancergene$enhancergene, "gain"] = T
plot_enhancers$geneset_gene = F
plot_enhancers[plot_enhancers$gene %in% geneset_genes, "geneset_gene"] = T



###### get K27ac plotting data 
K27ac_dba = dba.count(K27ac_dba, peaks=NULL, score=DBA_SCORE_READS_MINUS)
K27ac_enhancers = paste0(K27ac_dba$peaks[[1]]$Chr, ":", K27ac_dba$peaks[[1]]$Start, "-", K27ac_dba$peaks[[1]]$End)



################ do deseq normalization using NON-ADJUSTED DESEQ
K27ac_plotdata = K27ac_dba$binding[, -1:-3]
K27ac_plotdata = K27ac_plotdata[match(plot_enhancers$enhancer, K27ac_enhancers), match(plot_samples$Sample, colnames(K27ac_plotdata))]
K27ac_plotdata[K27ac_plotdata < 0] = 0
deseq2facs = deseq2facs[match(colnames(K27ac_plotdata), names(deseq2facs))]
K27ac_plotdata = t(t(K27ac_plotdata) / deseq2facs)
K27ac_plotdata = 1 + K27ac_plotdata
K27ac_plotdata = log2(K27ac_plotdata)



######## plot K27ac heatmap
plot_samples$color = "yellow"
plot_samples[plot_samples$EGroup=='Fluke', 'color'] = 'blue'
plot_samples[plot_samples$EGroup=='C4', 'color'] = 'red2'
plot_samples[plot_samples$EGroup=='Mixed', 'color'] = 'green3'
plot_samples[plot_samples$EGroup=='Normal', 'color'] = 'black'
plot_enhancers$color = 'white'
plot_enhancers[plot_enhancers$geneset_gene, 'color'] = 'black'
my_palette <- colorRampPalette(c("white", "darkgreen"))(n = 299)
col_breaks = c(seq(0,2,length=100),
  seq(2.01,4,length=100),
  seq(4.01,6,length=100))
heatmap_mat = heatmap.2(K27ac_plotdata, Colv = FALSE, Rowv = FALSE, dendrogram='none', density.info = "none", keysize = 1, denscol="none", 
                        rowsep=c(sum(plot_enhancers$gain)), sepcolor="black",
                        trace="none", scale='none', col=my_palette, breaks=col_breaks, ColSideColors=plot_samples$color, RowSideColors=plot_enhancers$color)





# plot tracks
heatmap_colnames = rownames(heatmap_mat$carpet)
zz_plot = samples_data[match(heatmap_colnames, samples_data$Sample),c("Sample", "Cluster", "Fluke", "AA", "Viral", "Location")]
zz_plot$ColID = 1:nrow(zz_plot)
zz_plot = pivot_longer(zz_plot, cols=Cluster | Fluke | AA | Viral | Location)
zz_plot$RowID = rep(c(5,4,3,2,1), nrow(zz_plot)/5)
zz_plot$'name-value' = paste0(zz_plot$name, ":", zz_plot$value)
zz_plot$color = "yellow"
zz_plot[zz_plot$`name-value`=="Cluster:Normal", 'color'] = 'black'
zz_plot[zz_plot$`name-value`=="Cluster:Normal_liver", 'color'] = 'brown'
zz_plot[zz_plot$`name-value`=="Cluster:1", 'color'] = 'blue'
zz_plot[zz_plot$`name-value`=="Cluster:2", 'color'] = 'orange'
zz_plot[zz_plot$`name-value`=="Cluster:3", 'color'] = 'green'
zz_plot[zz_plot$`name-value`=="Cluster:4", 'color'] = 'red'
zz_plot[zz_plot$`name-value`=="Cluster:N/A", 'color'] = 'grey'
zz_plot[zz_plot$`name-value`=="Fluke:Fluke-Neg", 'color'] = 'white'
zz_plot[zz_plot$`name-value`=="Fluke:Fluke-Pos", 'color'] = 'black'
zz_plot[zz_plot$`name-value`=="AA:No", 'color'] = 'white'
zz_plot[zz_plot$`name-value`=="AA:Yes", 'color'] = 'black'
zz_plot[zz_plot$`name-value`=="AA:N/A", 'color'] = 'grey'
zz_plot[zz_plot$`name-value`=="Viral:N/A", 'color'] = 'grey'
zz_plot[zz_plot$`name-value` %in% c("Viral:HBV", "Viral:HCV", "Viral:HBV+HCV"), 'color'] = 'black'
zz_plot[zz_plot$`name-value`=="Viral:Negative", 'color'] = 'white'
zz_plot[zz_plot$`name-value`=="Location:N/A", 'color'] = 'grey'
zz_plot[zz_plot$`name-value`=="Location:Intrahepatic", 'color'] = 'orange'
zz_plot[zz_plot$`name-value`=="Location:Perihilar", 'color'] = 'blue'
zz_plot[zz_plot$`name-value`=="Location:Distal", 'color'] = 'green'
zz_plot[zz_plot$`name-value`=="Location:Extrahepatic", 'color'] = 'red'
zz_plot[zz_plot$`name-value`=="Location:Cystic duct", 'color'] = 'black'
zz_plot[zz_plot$`name-value`=="Location:Liver", 'color'] = 'white'


ggplot(data=zz_plot, mapping = aes(x=ColID, y=RowID)) +
  geom_tile(aes(width = 1,height = 1 , fill = color)) +
  scale_fill_manual(breaks=c("black","blue","green", "grey", "orange", "red", "white"), values = c("black","blue","green3", "grey", "orange", "red2", "white")) +
  theme_void()



col_breaks = c(seq(-2.5,-1.01,length=100),
  seq(-1,1,length=100),
  seq(1.01,2.5,length=100))
heatmap_mat = heatmap.2(K27ac_plotdata, Colv = FALSE, Rowv = FALSE, dendrogram='none',  density.info = "none", keysize = 1, denscol="none", 
                        rowsep=c(sum(plot_enhancers$gain)), sepcolor="black",
                        trace="none", scale='row', col=my_palette, breaks=col_breaks, ColSideColors=plot_samples$color, RowSideColors=plot_enhancers$color)


```







### plot cell-line enhancers

```{r fig.width=12}

load(file="Rdata_cca_k27ac_3c_cells_diffbind.Rdata")
rm(K27ac_cells_dba, K27ac_cells_enhsup_tisspeaks_dba)

samples_info_cell = dba.show(K27ac_cells_tisspeaks_dba)
samples_info_cell$Sample = as.character(samples_info_cell$ID)
samples_info_cell$Tissue = as.character(samples_info_cell$Tissue)
samples_info_cell$Condition = as.character(samples_info_cell$Condition)
samples_info_cell[samples_info_cell$Tissue=="Normal", "Condition"] = "Normal"

####### get deseq2 normalization factors
# deseq2facs_cell = get_deseq2_normfacs(K27ac_cells_tisspeaks_dba, 
#               samps=samples_info_cell[samples_info_cell$Condition %in% c("Normal", "Fluke-Pos"),"Sample"])


# get factors based on raw counts, BUT SET NEGATIVES TO 0 FIRST!
zz_K27ac_dba = dba(DBA=K27ac_cells_tisspeaks_dba, mask=K27ac_cells_tisspeaks_dba$masks$Normal | K27ac_cells_tisspeaks_dba$masks$`Fluke-Pos`) 
zz_K27ac_dba = dba.count(zz_K27ac_dba, peaks=NULL, score=DBA_SCORE_READS_MINUS)
zz_K27ac_data = zz_K27ac_dba$binding[, -1:-3]
zz_K27ac_data[zz_K27ac_data < 0] = 0
deseq2facs_cell = estimateSizeFactorsForMatrix(zz_K27ac_data)
rm(zz_K27ac_dba, zz_K27ac_data)


###### order the data to plot
plot_samples_cell = samples_info_cell[, c("Sample", "Condition")]
plot_samples_cell = plot_samples_cell[plot_samples_cell$Condition %in% c("Fluke-Pos", "Normal"),]
plot_samples_cell = plot_samples_cell[order(factor(plot_samples_cell$Condition, levels=c("Fluke-Pos", "Normal"))),]
# make sure plot_enhancers is obtained from above chunk


###### get K27ac plotting data 
K27ac_cells_tisspeaks_dba = dba.count(K27ac_cells_tisspeaks_dba, peaks=NULL, score=DBA_SCORE_READS_MINUS)
K27ac_enhancers = paste0(K27ac_cells_tisspeaks_dba$peaks[[1]]$Chr, ":", K27ac_cells_tisspeaks_dba$peaks[[1]]$Start, "-", K27ac_cells_tisspeaks_dba$peaks[[1]]$End)




################ do deseq normalization 
K27ac_plotdata = K27ac_cells_tisspeaks_dba$binding[, -1:-3]
K27ac_plotdata = K27ac_plotdata[match(plot_enhancers$enhancer, K27ac_enhancers), match(plot_samples_cell$Sample, colnames(K27ac_plotdata))]
K27ac_plotdata[K27ac_plotdata < 0] = 0
deseq2facs_cell = deseq2facs_cell[match(colnames(K27ac_plotdata), names(deseq2facs_cell))]
K27ac_plotdata = t(t(K27ac_plotdata) / deseq2facs_cell)
K27ac_plotdata = 1 + K27ac_plotdata
K27ac_plotdata = log2(K27ac_plotdata)



######## plot K27ac heatmap
my_palette <- colorRampPalette(c("white", "darkgreen"))(n = 299)
col_breaks = c(seq(0,2,length=100),
  seq(2.01,4,length=100),
  seq(4.01,6,length=100))
plot_samples_cell$color = "yellow"
plot_samples_cell[plot_samples_cell$Condition=='Fluke-Pos', 'color'] = 'blue'
plot_samples_cell[plot_samples_cell$Condition=='Normal', 'color'] = 'black'
heatmap_mat = heatmap.2(K27ac_plotdata, Colv = FALSE, Rowv = FALSE, dendrogram='none', density.info = "none",  keysize = 1, denscol="none", 
                        rowsep=c(sum(plot_enhancers$gain)), sepcolor="black",
                        trace="none", scale='none', col=my_palette, breaks=col_breaks, ColSideColors=plot_samples_cell$color)


```









### plot tissue expression

```{r fig.width=12}
library(DESeq2)
library(RColorBrewer)


# read expression data: tissue counts. Ensure samples have same IDs as chipseq data!
expr_counts_tiss_data = read.table("expression/counts_matrix_tissue_CCA_HCC/counts_matrix-genes.txt", stringsAsFactors = F, header=T, check.names = F, sep="\t") # 26589 rows, all unique gene names.
rownames(expr_counts_tiss_data) = expr_counts_tiss_data$gene_name
expr_counts_tiss_data = expr_counts_tiss_data[, !colnames(expr_counts_tiss_data) %in% c("gene_id", "gene_name")]
expr_counts_tiss_data = round(expr_counts_tiss_data)

# get samples to plot
plot_samples = samples_data[samples_data$EGroup!="Normal_liver",]
plot_samples = plot_samples[order(factor(plot_samples$EGroup, levels=c("Fluke", "C4", "Mixed", "Normal"))),]
plot_samples = plot_samples[plot_samples$Sample %in% colnames(expr_counts_tiss_data),]

# re-order samples in expression data
expr_counts_tiss_data = expr_counts_tiss_data[, match(plot_samples$Sample, colnames(expr_counts_tiss_data))]

# get deseq normalization factors
deseq2facs_expr = estimateSizeFactorsForMatrix(expr_counts_tiss_data)

# do deseq normalization
deseq2facs_expr = deseq2facs_expr[match(colnames(expr_counts_tiss_data), names(deseq2facs_expr))]
expr_counts_tiss_data = t(t(expr_counts_tiss_data) / deseq2facs_expr)
expr_counts_tiss_data = 1 + expr_counts_tiss_data
expr_counts_tiss_data = log2(expr_counts_tiss_data)

# re-order genes in expression data. get plot_enhancers from chunk above!!!!
expr_counts_tiss_data = expr_counts_tiss_data[match(plot_enhancers$gene, rownames(expr_counts_tiss_data)),]



##### plot heatmap
plot_samples$color = "yellow"
plot_samples[plot_samples$EGroup=='Fluke', 'color'] = 'blue'
plot_samples[plot_samples$EGroup=='C4', 'color'] = 'red2'
plot_samples[plot_samples$EGroup=='Mixed', 'color'] = 'green3'
plot_samples[plot_samples$EGroup=='Normal', 'color'] = 'black'

# use blue - white - red
my_palette <- colorRampPalette(c("blue", "white", "red"))(n = 299)
# no row scaling
col_breaks = c(seq(0,4,length=100),
  seq(4.01,9,length=100),
  seq(9.01,14,length=100))
heatmap_mat = heatmap.2(expr_counts_tiss_data,
                 trace="none", scale="none", dendrogram="none", Rowv=NULL, Colv=NULL, density.info = "none", keysize = 1, denscol="none", 
                        rowsep=c(sum(plot_enhancers$gain)), sepcolor="black",
                 col=my_palette, breaks=col_breaks, ColSideColors=plot_samples$color)

# row scaling
col_breaks = c(seq(-2.5,-1.01,length=100),
  seq(-1,1,length=100),
  seq(1.01,2.5,length=100))
heatmap_mat = heatmap.2(expr_counts_tiss_data,
                 trace="none", scale="row", dendrogram="none", Rowv=NULL, Colv=NULL, density.info = "none", keysize = 1, denscol="none", 
                        rowsep=c(sum(plot_enhancers$gain)), sepcolor="black",
                 col=my_palette, breaks=col_breaks,  ColSideColors=plot_samples$color)





# plot tracks
heatmap_colnames = rownames(heatmap_mat$carpet)
zz_plot = samples_data[match(heatmap_colnames, samples_data$Sample),c("Sample", "Cluster", "Fluke", "AA", "Viral", "Location")]
zz_plot$ColID = 1:nrow(zz_plot)
zz_plot = pivot_longer(zz_plot, cols=Cluster | Fluke | AA | Viral | Location)
zz_plot$RowID = rep(c(5,4,3,2,1), nrow(zz_plot)/5)
zz_plot$'name-value' = paste0(zz_plot$name, ":", zz_plot$value)
zz_plot$color = "yellow"
zz_plot[zz_plot$`name-value`=="Cluster:Normal", 'color'] = 'black'
zz_plot[zz_plot$`name-value`=="Cluster:Normal_liver", 'color'] = 'brown'
zz_plot[zz_plot$`name-value`=="Cluster:1", 'color'] = 'blue'
zz_plot[zz_plot$`name-value`=="Cluster:2", 'color'] = 'orange'
zz_plot[zz_plot$`name-value`=="Cluster:3", 'color'] = 'green'
zz_plot[zz_plot$`name-value`=="Cluster:4", 'color'] = 'red'
zz_plot[zz_plot$`name-value`=="Cluster:N/A", 'color'] = 'grey'
zz_plot[zz_plot$`name-value`=="Fluke:Fluke-Neg", 'color'] = 'white'
zz_plot[zz_plot$`name-value`=="Fluke:Fluke-Pos", 'color'] = 'black'
zz_plot[zz_plot$`name-value`=="AA:No", 'color'] = 'white'
zz_plot[zz_plot$`name-value`=="AA:Yes", 'color'] = 'black'
zz_plot[zz_plot$`name-value`=="AA:N/A", 'color'] = 'grey'
zz_plot[zz_plot$`name-value`=="Viral:N/A", 'color'] = 'grey'
zz_plot[zz_plot$`name-value` %in% c("Viral:HBV", "Viral:HCV", "Viral:HBV+HCV"), 'color'] = 'black'
zz_plot[zz_plot$`name-value`=="Viral:Negative", 'color'] = 'white'
zz_plot[zz_plot$`name-value`=="Location:N/A", 'color'] = 'grey'
zz_plot[zz_plot$`name-value`=="Location:Intrahepatic", 'color'] = 'orange'
zz_plot[zz_plot$`name-value`=="Location:Perihilar", 'color'] = 'blue'
zz_plot[zz_plot$`name-value`=="Location:Distal", 'color'] = 'green'
zz_plot[zz_plot$`name-value`=="Location:Extrahepatic", 'color'] = 'red'
zz_plot[zz_plot$`name-value`=="Location:Cystic duct", 'color'] = 'black'
zz_plot[zz_plot$`name-value`=="Location:Liver", 'color'] = 'white'

```








### plot cell lines expression

```{r fig.width=12}
library(DESeq2)
library(RColorBrewer)


# read expression data: cellline counts. Ensure samples have same IDs as chipseq data!
expr_counts_cell_data = read.table("expression/counts_matrix_celllines/counts_matrix-genes.txt", stringsAsFactors = F, header=T, check.names = F, sep="\t") # 26589 rows, all unique gene names.
rownames(expr_counts_cell_data) = expr_counts_cell_data$gene_name
expr_counts_cell_data = expr_counts_cell_data[, !colnames(expr_counts_cell_data) %in% c("gene_id", "gene_name")]
expr_counts_cell_data = round(expr_counts_cell_data)
expr_counts_cell_data = expr_counts_cell_data[, grep("H69-KO|S5-EV|S5-MUT|S5-WT", colnames(expr_counts_cell_data), invert = T)]
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="Egi-1-1"] = "EGI1_R0"
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="Egi-1-2"] = "EGI1_R1"
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="H69-1"] = "H69_R0"
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="H69-2"] = "H69_R1"
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="HUCCT"] = "HUCCT_R1"
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="KKU100"] = "KKU100_R1"
colnames(expr_counts_cell_data)[colnames(expr_counts_cell_data)=="M213"] = "M214"

# re-order samples in expression data. get plot_samples_cell from above. 
plot_samples_cell_expr = plot_samples_cell[plot_samples_cell$Sample %in% colnames(expr_counts_cell_data),]
expr_counts_cell_data = expr_counts_cell_data[, match(plot_samples_cell_expr$Sample, colnames(expr_counts_cell_data))]

# get deseq normalization factors
deseq2facs_cell_expr = estimateSizeFactorsForMatrix(expr_counts_cell_data)

# do deseq normalization
deseq2facs_cell_expr = deseq2facs_cell_expr[match(colnames(expr_counts_cell_data), names(deseq2facs_cell_expr))]
expr_counts_cell_data = t(t(expr_counts_cell_data) / deseq2facs_cell_expr)
expr_counts_cell_data = 1 + expr_counts_cell_data
expr_counts_cell_data = log2(expr_counts_cell_data)

# re-order genes in expression data. get plot_enhancers from chunk above!!!!
expr_counts_cell_data = expr_counts_cell_data[match(plot_enhancers$gene, rownames(expr_counts_cell_data)),]



# plot heatmap
my_palette <- colorRampPalette(c("blue", "white", "red"))(n = 299)
col_breaks = c(seq(0,3,length=100),
  seq(4.01,9,length=100),
  seq(9.01,14,length=100))
heatmap_mat = heatmap.2(expr_counts_cell_data,
                 trace="none", scale="none", dendrogram="none", Rowv=NULL, Colv=NULL, density.info = "none", keysize = 1, denscol="none", 
                        rowsep=c(sum(plot_enhancers$gain)), sepcolor="black",
                 col=my_palette, breaks=col_breaks, ColSideColors=plot_samples_cell_expr$color)
  
col_breaks = c(seq(-2.5,-1.01,length=100),
  seq(-1,1,length=100),
  seq(1.01,2.5,length=100))
heatmap_mat = heatmap.2(expr_counts_cell_data,
                 trace="none", scale="row", dendrogram="none", Rowv=NULL, Colv=NULL, density.info = "none", keysize = 1, denscol="none", 
                        rowsep=c(sum(plot_enhancers$gain)), sepcolor="black",
                 col=my_palette, breaks=col_breaks,  ColSideColors=plot_samples_cell_expr$color)
  

```










# investigate the estrogen-response genes gained


```{r include=F}

# options(scipen=0)
rm(list=ls())
library(DiffBind, warn.conflicts=F, quietly=T)
library(beeswarm)
library(msigdbr)
library(fgsea)
library(gplots)
library(clusterProfiler)
library(biomaRt)
source('analysis_cca_enhancers_functions.R')


# # read tissue enhancer data
# load(file="Rdata_cca_k27ac_0_prelim_final.Rdata")
# rm(K27ac_enhsup_dba, K27ac_origPeaks, superenhancers)


# read diffbind results
load(file="Rdata_cca_k27ac_3_cells_diffbind.Rdata")
rm(K27ac_cells_dba, K27ac_cells_enhsup_tisspeaks_dba)
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts1.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts2.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts3.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts1_diffexpr.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts2_diffexpr.Rdata")
load(file="Rdata_cca_k27ac_2b.2_diffbind_contrasts3_diffexpr.Rdata")


#  load multicomp results
load("Rdata_cca_k27ac_3a_somatic_enhancers.Rdata")
load("Rdata_cca_k27ac_3b_somatic_enhancers_groupspec.Rdata")


# read mapping
target_map = read.table("target_gene_mapping/20190628_enhancer_target_map_noTAD_CPLW_rho.3_liberal.txt", header=T, stringsAsFactors = F) 
target_map = target_map[!is.na(target_map$gene),c("enhancer", "gene")]
target_map = target_map[!duplicated(target_map),]




# read mSigDB genesets
msigdb_H = msigdbr(species = "Homo sapiens", category = "H") # 50 genesets
msigdb_H_list = msigdb_H %>% split(x = .$gene_symbol, f = .$gs_name)
msigdb_H_term2gene = data.frame(TERM=msigdb_H$gs_name, GENE=msigdb_H$gene_symbol, stringsAsFactors = F)
msigdb_H_term2gene = rbind(msigdb_H_term2gene, data.frame(TERM="ALL_GENES", GENE=unique(na.omit(target_map$gene)), stringsAsFactors = F)) # clusterProfiler considers universe = genes in genesets


# read biomart gene descriptions
biomart_ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")


get_genesets = function(gene, pathways=c("H","R")) {
  ann = NULL
  if ("H" %in% pathways) {
    ann = c(ann, msigdb_H_term2gene[msigdb_H_term2gene$GENE==gene,"TERM"])
  }
  if ("R" %in% pathways) {
    ann = c(ann, msigdb_R_term2gene[msigdb_R_term2gene$GENE==gene,"TERM"])
  }
  ann = unique(ann)
  ann = ann[ann!="ALL_GENES"]
  ann = gsub("HALLMARK", "H", ann)
  ann = gsub("REACTOME", "R", ann)
  ann = sapply(ann, function(x){substr(x, 1, 30)})
  paste0(ann, collapse=",")
}
get_gene_descr = function(gene){
  descr = getBM(attributes=c('description'), filters='external_gene_name', values=gene, mart=biomart_ensembl)[1,1]
  if (is.na(descr)) { "" 
  } else {  strsplit(descr, split=" [", fixed=T)[[1]][1] }
}


samples_data = read.table("2021.08.22 samples_info.txt", stringsAsFactors = F, header=T, sep="\t", check.names = F)


### put results here
# estrogen_investigation$non_groupspec$sig$genesets$Fluke$estroearly = enrichment of genesets in non-groupspecific early-estrogen genes gained by significance test, for Fluke group
# estrogen_investigation$non_groupspec$sig$mtorc_res = MTORC intersection with non-groupspecific estrogen genes gained by significance test
# estrogen_investigation$non_groupspec$gsea$genesets$Fluke$estroearly = enrichment of genesets in non-groupspecific early-estrogen genes gained by GSEA core enrichment
# estrogen_investigation$non_groupspec$gsea$mtorc_res = MTORC intersection with non-groupspecific estrogen genes gained by GSEA core enrichment
# estrogen_investigation$groupsec = same, using gained group-specific estrogen genes 
estrogen_investigation = list()


msigdb_estro = union(msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_EARLY, msigdb_H_list$HALLMARK_ESTROGEN_RESPONSE_LATE)
msigdb_mtorc = union(msigdb_H_list$HALLMARK_MTORC1_SIGNALING, msigdb_H_list$HALLMARK_PI3K_AKT_MTOR_SIGNALING)
msigdb_H_term2gene = rbind(msigdb_H_term2gene, data.frame(TERM=rep("YONG_MTORC", length(msigdb_mtorc)), GENE=msigdb_mtorc))
```


## get enrichment of gained estrogen genes

#### non-group-specific estrogen genes, gained by significance test

```{r fig.width=8}

# store overlap results with HALLMARK_MTORC1_SIGNALING, HALLMARK_PI3K_AKT_MTOR_SIGNALING here
mtorc_res = data.frame(matrix(ncol = 5, nrow = 3))
colnames(mtorc_res) = c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")
rownames(mtorc_res) = c("Fluke", "C4", "Mixed")
estrogen_investigation$nongroupspec$sig$mtorc_res = mtorc_res

#### Fluke group
# get gained estrogen genes
Fluke_estro = intersect(somatic_enhancers$Tumor_EGroupFluke_vs_Normal$gain_exp_enhgain$gene, msigdb_estro)


# what are gained estrogen genes enriched in?
Fluke_estro_enrich = attributes(enricher(gene=Fluke_estro, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene, pvalueCutoff=1))$result 
estrogen_investigation$groupspec$sig$genesets$Fluke$estro = Fluke_estro_enrich

estrogen_investigation$nongroupspec$sig$mtorc_res["Fluke", c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")] = 
  c(length(Fluke_estro), Fluke_estro_enrich['YONG_MTORC', c("Count", "pvalue", "p.adjust", "geneID")])



#### C4 group
# get gained estrogen genes
C4_estro = intersect(somatic_enhancers$Tumor_EGroupC4_vs_Normal$gain_exp_enhgain$gene, msigdb_estro)


# what are gained estrogen genes enriched in?
C4_estro_enrich = attributes(enricher(gene=C4_estro, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene, pvalueCutoff=1))$result 
estrogen_investigation$groupspec$sig$genesets$C4$estro = C4_estro_enrich

estrogen_investigation$nongroupspec$sig$mtorc_res["C4", c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")] = 
  c(length(C4_estro), C4_estro_enrich['YONG_MTORC', c("Count", "pvalue", "p.adjust", "geneID")])


#### Mixed group
# get gained estrogen genes
Mixed_estro = intersect(somatic_enhancers$Tumor_EGroupMixed_vs_Normal$gain_exp_enhgain$gene, msigdb_estro)


# what are gained estrogen genes enriched in?
Mixed_estro_enrich = attributes(enricher(gene=Mixed_estro, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene, pvalueCutoff=1))$result 
estrogen_investigation$groupspec$sig$genesets$Mixed$estro = Mixed_estro_enrich

estrogen_investigation$nongroupspec$sig$mtorc_res["Mixed", c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")] = 
  c(length(Mixed_estro), Mixed_estro_enrich['YONG_MTORC', c("Count", "pvalue", "p.adjust", "geneID")])






```





#### non-group-specific estrogen genes, gained by GSEA core enrichment
```{r}


# store overlap results with HALLMARK_MTORC1_SIGNALING, HALLMARK_PI3K_AKT_MTOR_SIGNALING here
mtorc_res = data.frame(matrix(ncol = 5, nrow = 3))
colnames(mtorc_res) = c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")
rownames(mtorc_res) = c("Fluke", "C4", "Mixed")
estrogen_investigation$nongroupspec$gsea$mtorc_res = mtorc_res


#### Fluke group
# get gsea core enrichment genes
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts1, K27ac_dba2=K27ac_dba_deseq_contrasts1, contrasts=c(1,1), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts1, diffexprlist2=diffexpr_list_contrasts1, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=2, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)

Fluke_gsea = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000, pvalueCutoff=1)
Fluke_gsea = attributes(Fluke_gsea)$result[attributes(Fluke_gsea)$result$enrichmentScore>=0,]

Fluke_estro = unique(c(strsplit(split='/', Fluke_gsea['HALLMARK_ESTROGEN_RESPONSE_EARLY', 'core_enrichment'])[[1]], 
                       strsplit(split='/', Fluke_gsea['HALLMARK_ESTROGEN_RESPONSE_LATE', 'core_enrichment'])[[1]]))


# what are gained estrogen genes enriched in?
Fluke_estro_enrich = attributes(enricher(gene=Fluke_estro, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene, pvalueCutoff=1))$result 
estrogen_investigation$nongroupspec$gsea$genesets$Fluke$estro = Fluke_estro_enrich

# get overlap with HALLMARK_MTORC1_SIGNALING, HALLMARK_PI3K_AKT_MTOR_SIGNALING
estrogen_investigation$nongroupspec$gsea$mtorc_res["Fluke", c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")] = 
  c(length(Fluke_estro), Fluke_estro_enrich['YONG_MTORC', c("Count", "pvalue", "p.adjust", "geneID")])



#### C4 group
# get gsea core enrichment genes
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts2, K27ac_dba2=K27ac_dba_deseq_contrasts2, contrasts=c(1,1), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts2, diffexprlist2=diffexpr_list_contrasts2, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)

C4_gsea = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000, pvalueCutoff=1)
C4_gsea = attributes(C4_gsea)$result[attributes(C4_gsea)$result$enrichmentScore>=0,]

C4_estro = unique(c(strsplit(split='/', C4_gsea['HALLMARK_ESTROGEN_RESPONSE_EARLY', 'core_enrichment'])[[1]], 
                       strsplit(split='/', C4_gsea['HALLMARK_ESTROGEN_RESPONSE_LATE', 'core_enrichment'])[[1]]))


# what are gained estrogen genes enriched in?
C4_estro_enrich = attributes(enricher(gene=C4_estro, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene, pvalueCutoff=1))$result 
estrogen_investigation$nongroupspec$gsea$genesets$C4$estro = C4_estro_enrich

# get overlap with HALLMARK_MTORC1_SIGNALING, HALLMARK_PI3K_AKT_MTOR_SIGNALING
estrogen_investigation$nongroupspec$gsea$mtorc_res["C4", c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")] = 
  c(length(C4_estro), C4_estro_enrich['YONG_MTORC', c("Count", "pvalue", "p.adjust", "geneID")])




#### Mixed group
# get gsea core enrichment genes
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts3, K27ac_dba2=K27ac_dba_deseq_contrasts3, contrasts=c(1,1), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts3, diffexprlist2=diffexpr_list_contrasts3, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)

Mixed_gsea = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000, pvalueCutoff=1)
Mixed_gsea = attributes(Mixed_gsea)$result[attributes(Mixed_gsea)$result$enrichmentScore>=0,]

Mixed_estro = unique(c(strsplit(split='/', Mixed_gsea['HALLMARK_ESTROGEN_RESPONSE_EARLY', 'core_enrichment'])[[1]], 
                       strsplit(split='/', Mixed_gsea['HALLMARK_ESTROGEN_RESPONSE_LATE', 'core_enrichment'])[[1]]))


# what are gained estrogen genes enriched in?
Mixed_estro_enrich = attributes(enricher(gene=Mixed_estro, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene, pvalueCutoff=1))$result 
estrogen_investigation$nongroupspec$gsea$genesets$Mixed$estro = Mixed_estro_enrich

# get overlap with HALLMARK_MTORC1_SIGNALING, HALLMARK_PI3K_AKT_MTOR_SIGNALING
estrogen_investigation$nongroupspec$gsea$mtorc_res["Mixed", c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")] = 
  c(length(Mixed_estro), Mixed_estro_enrich['YONG_MTORC', c("Count", "pvalue", "p.adjust", "geneID")])







```


#### group-specific estrogen genes, gained by significance test

```{r fig.width=8}

# store overlap results with HALLMARK_MTORC1_SIGNALING, HALLMARK_PI3K_AKT_MTOR_SIGNALING here
mtorc_res = data.frame(matrix(ncol = 5, nrow = 3))
colnames(mtorc_res) = c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")
rownames(mtorc_res) = c("Fluke", "C4", "Mixed")
estrogen_investigation$groupspec$sig$mtorc_res = mtorc_res

#### Fluke group
# get gained estrogen genes
Fluke_estro = intersect(intersect(somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain$gene,
                                                  somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain$gene),
                                        msigdb_estro)


# what are gained estrogen genes enriched in?
Fluke_estro_enrich = attributes(enricher(gene=Fluke_estro, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene, pvalueCutoff=1))$result 
estrogen_investigation$groupspec$sig$genesets$Fluke$estro = Fluke_estro_enrich

estrogen_investigation$groupspec$sig$mtorc_res["Fluke", c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")] = 
  c(length(Fluke_estro), Fluke_estro_enrich['YONG_MTORC', c("Count", "pvalue", "p.adjust", "geneID")])



#### C4 group
# get gained estrogen genes
C4_estro = intersect(intersect(somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene,
                                                  somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain$gene),
                                        msigdb_estro)


# what are gained estrogen genes enriched in?
C4_estro_enrich = attributes(enricher(gene=C4_estro, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene, pvalueCutoff=1))$result 
estrogen_investigation$groupspec$sig$genesets$C4$estro = C4_estro_enrich

estrogen_investigation$groupspec$sig$mtorc_res["C4", c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")] = 
  c(length(C4_estro), C4_estro_enrich['YONG_MTORC', c("Count", "pvalue", "p.adjust", "geneID")])




#### Mixed group
# get gained estrogen genes
Mixed_estro = intersect(intersect(somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain$gene,
                                                  somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain$gene),
                                        msigdb_estro)


# what are gained estrogen genes enriched in?
Mixed_estro_enrich = attributes(enricher(gene=Mixed_estro, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene, pvalueCutoff=1))$result 
estrogen_investigation$groupspec$sig$genesets$Mixed$estro = Mixed_estro_enrich

estrogen_investigation$groupspec$sig$mtorc_res["Mixed", c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")] = 
  c(length(Mixed_estro), Mixed_estro_enrich['YONG_MTORC', c("Count", "pvalue", "p.adjust", "geneID")])


```




#### group-specific estrogen genes, gained by GSEA core enrichment
```{r}

# store overlap results with HALLMARK_MTORC1_SIGNALING, HALLMARK_PI3K_AKT_MTOR_SIGNALING here
mtorc_res = data.frame(matrix(ncol = 5, nrow = 3))
colnames(mtorc_res) = c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")
rownames(mtorc_res) = c("Fluke", "C4", "Mixed")
estrogen_investigation$groupspec$gsea$mtorc_res = mtorc_res


#### Fluke group
# get gsea core enrichment genes
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts1, K27ac_dba2=K27ac_dba_deseq_contrasts1, contrasts=c(1,3,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts1, diffexprlist2=diffexpr_list_contrasts1, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=2, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)

Fluke_gsea = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000, pvalueCutoff=1)
Fluke_gsea = attributes(Fluke_gsea)$result[attributes(Fluke_gsea)$result$enrichmentScore>=0,]

Fluke_estro = unique(c(strsplit(split='/', Fluke_gsea['HALLMARK_ESTROGEN_RESPONSE_EARLY', 'core_enrichment'])[[1]], 
                       strsplit(split='/', Fluke_gsea['HALLMARK_ESTROGEN_RESPONSE_LATE', 'core_enrichment'])[[1]]))


# what are gained estrogen genes enriched in?
Fluke_estro_enrich = attributes(enricher(gene=Fluke_estro, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene, pvalueCutoff=1))$result 
estrogen_investigation$groupspec$gsea$genesets$Fluke$estro = Fluke_estro_enrich

# get overlap with HALLMARK_MTORC1_SIGNALING, HALLMARK_PI3K_AKT_MTOR_SIGNALING
estrogen_investigation$groupspec$gsea$mtorc_res["Fluke", c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")] = 
  c(length(Fluke_estro), Fluke_estro_enrich['YONG_MTORC', c("Count", "pvalue", "p.adjust", "geneID")])



#### C4 group
# get gsea core enrichment genes
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts2, K27ac_dba2=K27ac_dba_deseq_contrasts2, contrasts=c(1,3,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts2, diffexprlist2=diffexpr_list_contrasts2, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)

C4_gsea = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000, pvalueCutoff=1)
C4_gsea = attributes(C4_gsea)$result[attributes(C4_gsea)$result$enrichmentScore>=0,]

C4_estro = unique(c(strsplit(split='/', C4_gsea['HALLMARK_ESTROGEN_RESPONSE_EARLY', 'core_enrichment'])[[1]], 
                       strsplit(split='/', C4_gsea['HALLMARK_ESTROGEN_RESPONSE_LATE', 'core_enrichment'])[[1]]))


# what are gained estrogen genes enriched in?
C4_estro_enrich = attributes(enricher(gene=C4_estro, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene, pvalueCutoff=1))$result 
estrogen_investigation$groupspec$gsea$genesets$C4$estro = C4_estro_enrich

# get overlap with HALLMARK_MTORC1_SIGNALING, HALLMARK_PI3K_AKT_MTOR_SIGNALING
estrogen_investigation$groupspec$gsea$mtorc_res["C4", c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")] = 
  c(length(C4_estro), C4_estro_enrich['YONG_MTORC', c("Count", "pvalue", "p.adjust", "geneID")])




#### Mixed group
# get gsea core enrichment genes
gsea_genelist_CellFilt2 = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts3, K27ac_dba2=K27ac_dba_deseq_contrasts3, contrasts=c(1,3,4), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts3, diffexprlist2=diffexpr_list_contrasts3, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=2, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
gsea_genelist_CellFilt2 = sort(gsea_genelist_CellFilt2, decreasing = T)

Mixed_gsea = GSEA(geneList=gsea_genelist_CellFilt2, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene, seed=12345, nPerm=1000, pvalueCutoff=1)
Mixed_gsea = attributes(Mixed_gsea)$result[attributes(Mixed_gsea)$result$enrichmentScore>=0,]

Mixed_estro = unique(c(strsplit(split='/', Mixed_gsea['HALLMARK_ESTROGEN_RESPONSE_EARLY', 'core_enrichment'])[[1]], 
                       strsplit(split='/', Mixed_gsea['HALLMARK_ESTROGEN_RESPONSE_LATE', 'core_enrichment'])[[1]]))


# what are gained estrogen genes enriched in?
Mixed_estro_enrich = attributes(enricher(gene=Mixed_estro, universe=unique(target_map$gene), TERM2GENE=msigdb_H_term2gene, pvalueCutoff=1))$result 
estrogen_investigation$groupspec$gsea$genesets$Mixed$estro = Mixed_estro_enrich

# get overlap with HALLMARK_MTORC1_SIGNALING, HALLMARK_PI3K_AKT_MTOR_SIGNALING
estrogen_investigation$groupspec$gsea$mtorc_res["Mixed", c("estro", "estro_mtorc", "estro_mtorc_p", "estro_mtorc_q", "estro_mtorc_genes")] = 
  c(length(Mixed_estro), Mixed_estro_enrich['YONG_MTORC', c("Count", "pvalue", "p.adjust", "geneID")])



```



```{r}
print("*** Non Group-specific ***")
print("Enhancers gained by sig test")
print(t(estrogen_investigation$nongroupspec$sig$mtorc_res))

print("")
print("Enhancers gained by gsea core enrich")
print(t(estrogen_investigation$nongroupspec$gsea$mtorc_res))

```




```{r}
print("*** Group-specific ***")
print("Enhancers gained by sig test")
print(t(estrogen_investigation$groupspec$sig$mtorc_res))

print("")
print("Enhancers gained by gsea core enrich")
print(t(estrogen_investigation$groupspec$gsea$mtorc_res))

```


## for group-specific estrogen enhancers (by sig test), check overlap with each geneset
```{r}

zz_msig = split(msigdb_H_term2gene, msigdb_H_term2gene$TERM)
zz_msig$ALL_GENES  = NULL
zz_msig$HALLMARK_ESTROGEN_RESPONSE_EARLY = NULL
zz_msig$HALLMARK_ESTROGEN_RESPONSE_LATE = NULL
zz_msig$YONG_MTORC  = NULL
```

#### Fluke
```{r fig.width=8}

gain_enhancergene1 = somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene1$enhancergene = paste0(gain_enhancergene1$enhancer, '|', gain_enhancergene1$gene)
gain_enhancergene1 = gain_enhancergene1[!duplicated(gain_enhancergene1$enhancergene),]
gain_enhancergene2 = somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene2$enhancergene = paste0(gain_enhancergene2$enhancer, '|', gain_enhancergene2$gene)
gain_enhancergene2 = gain_enhancergene2[!duplicated(gain_enhancergene2$enhancergene),]
gain_enhancergene = gain_enhancergene1[gain_enhancergene1$enhancergene %in% gain_enhancergene2$enhancergene,]

# calculate overlap of gained enhancers with estrogen & mtorc genesets
Fluke_genesets_intersection = lapply(zz_msig, function(x) { 
  zz_estro = gain_enhancergene[gain_enhancergene$gene %in% msigdb_estro, ]
  zz_estro = gain_enhancergene[gain_enhancergene$enhancer %in% zz_estro$enhancer, ] # take ALL rows of enhancers with any estrogen genes
  zz_nonestro = gain_enhancergene[!gain_enhancergene$enhancer %in% zz_estro$enhancer, ] # get enhancers that don't have ANY estrogen genes
  zz_geneset = gain_enhancergene[gain_enhancergene$gene %in% x$GENE, ]
  zz_geneset = gain_enhancergene[gain_enhancergene$enhancer %in% zz_geneset$enhancer, ] # take ALL rows of enhancers with any geneset genes
  zz_nongeneset = gain_enhancergene[!gain_enhancergene$enhancer %in% zz_geneset$enhancer, ] # get enhancers that don't have ANY geneset genes
  
  zz_estro_geneset = zz_estro[zz_estro$gene %in% x$GENE, ]
  zz_estro_geneset = zz_estro[zz_estro$enhancer %in% zz_estro_geneset$enhancer, ] 
  zz_estro_nongeneset = zz_estro[!zz_estro$enhancer %in% zz_estro_geneset$enhancer, ]
  zz_nonestro_geneset = zz_nonestro[zz_nonestro$gene %in% x$GENE, ]
  zz_nonestro_geneset = zz_nonestro[zz_nonestro$enhancer %in% zz_nonestro_geneset$enhancer, ]
  zz_nonestro_nongeneset = zz_nonestro[!zz_nonestro$enhancer %in% zz_nonestro_geneset$enhancer, ]
  
  zz_fisher = fisher.test(matrix(c(length(unique(zz_estro_geneset$enhancer)), length(unique(zz_nonestro_geneset$enhancer)),
                                   length(unique(zz_estro_nongeneset$enhancer)), length(unique(zz_nonestro_nongeneset$enhancer))),nrow=2))

  return (data.frame(All_gain=length(unique(gain_enhancergene$enhancer)),
                     Estro=length(unique(zz_estro$enhancer)), NonEstro=length(unique(zz_nonestro$enhancer)),
                     Geneset=length(unique(zz_geneset$enhancer)), NonGeneset=length(unique(zz_nongeneset$enhancer)),
                     EstroGeneset=length(unique(zz_estro_geneset$enhancer)), EstroNonGeneset=length(unique(zz_estro_nongeneset$enhancer)),
                     NonEstroGeneset=length(unique(zz_nonestro_geneset$enhancer)), NonEstroNonGeneset=length(unique(zz_nonestro_nongeneset$enhancer)),
                     FisherP = zz_fisher$p.value))
})
Fluke_genesets_intersection = do.call(rbind, Fluke_genesets_intersection)
Fluke_genesets_intersection$FDR = p.adjust(Fluke_genesets_intersection$FisherP, method='BH')
Fluke_genesets_intersection = Fluke_genesets_intersection[order(Fluke_genesets_intersection$FDR),]
Fluke_genesets_intersection
```







#### C4
```{r}

gain_enhancergene1 = somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene1$enhancergene = paste0(gain_enhancergene1$enhancer, '|', gain_enhancergene1$gene)
gain_enhancergene1 = gain_enhancergene1[!duplicated(gain_enhancergene1$enhancergene),]
gain_enhancergene2 = somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene2$enhancergene = paste0(gain_enhancergene2$enhancer, '|', gain_enhancergene2$gene)
gain_enhancergene2 = gain_enhancergene2[!duplicated(gain_enhancergene2$enhancergene),]
gain_enhancergene = gain_enhancergene1[gain_enhancergene1$enhancergene %in% gain_enhancergene2$enhancergene,]

# calculate overlap of gained enhancers with estrogen & mtorc genesets
C4_genesets_intersection = lapply(zz_msig, function(x) { 
  zz_estro = gain_enhancergene[gain_enhancergene$gene %in% msigdb_estro, ]
  zz_estro = gain_enhancergene[gain_enhancergene$enhancer %in% zz_estro$enhancer, ] # take ALL rows of enhancers with any estrogen genes
  zz_nonestro = gain_enhancergene[!gain_enhancergene$enhancer %in% zz_estro$enhancer, ] # get enhancers that don't have ANY estrogen genes
  zz_geneset = gain_enhancergene[gain_enhancergene$gene %in% x$GENE, ]
  zz_geneset = gain_enhancergene[gain_enhancergene$enhancer %in% zz_geneset$enhancer, ] # take ALL rows of enhancers with any geneset genes
  zz_nongeneset = gain_enhancergene[!gain_enhancergene$enhancer %in% zz_geneset$enhancer, ] # get enhancers that don't have ANY geneset genes
  
  zz_estro_geneset = zz_estro[zz_estro$gene %in% x$GENE, ]
  zz_estro_geneset = zz_estro[zz_estro$enhancer %in% zz_estro_geneset$enhancer, ] 
  zz_estro_nongeneset = zz_estro[!zz_estro$enhancer %in% zz_estro_geneset$enhancer, ]
  zz_nonestro_geneset = zz_nonestro[zz_nonestro$gene %in% x$GENE, ]
  zz_nonestro_geneset = zz_nonestro[zz_nonestro$enhancer %in% zz_nonestro_geneset$enhancer, ]
  zz_nonestro_nongeneset = zz_nonestro[!zz_nonestro$enhancer %in% zz_nonestro_geneset$enhancer, ]
  
  zz_fisher = fisher.test(matrix(c(length(unique(zz_estro_geneset$enhancer)), length(unique(zz_nonestro_geneset$enhancer)),
                                   length(unique(zz_estro_nongeneset$enhancer)), length(unique(zz_nonestro_nongeneset$enhancer))),nrow=2))

  return (data.frame(All_gain=length(unique(gain_enhancergene$enhancer)),
                     Estro=length(unique(zz_estro$enhancer)), NonEstro=length(unique(zz_nonestro$enhancer)),
                     Geneset=length(unique(zz_geneset$enhancer)), NonGeneset=length(unique(zz_nongeneset$enhancer)),
                     EstroGeneset=length(unique(zz_estro_geneset$enhancer)), EstroNonGeneset=length(unique(zz_estro_nongeneset$enhancer)),
                     NonEstroGeneset=length(unique(zz_nonestro_geneset$enhancer)), NonEstroNonGeneset=length(unique(zz_nonestro_nongeneset$enhancer)),
                     FisherP = zz_fisher$p.value))
})
C4_genesets_intersection = do.call(rbind, C4_genesets_intersection)
C4_genesets_intersection$FDR = p.adjust(C4_genesets_intersection$FisherP, method='BH')
C4_genesets_intersection = C4_genesets_intersection[order(C4_genesets_intersection$FDR),]
C4_genesets_intersection
```




#### Mixed
```{r}

gain_enhancergene1 = somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene1$enhancergene = paste0(gain_enhancergene1$enhancer, '|', gain_enhancergene1$gene)
gain_enhancergene1 = gain_enhancergene1[!duplicated(gain_enhancergene1$enhancergene),]
gain_enhancergene2 = somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene2$enhancergene = paste0(gain_enhancergene2$enhancer, '|', gain_enhancergene2$gene)
gain_enhancergene2 = gain_enhancergene2[!duplicated(gain_enhancergene2$enhancergene),]
gain_enhancergene = gain_enhancergene1[gain_enhancergene1$enhancergene %in% gain_enhancergene2$enhancergene,]

# calculate overlap of gained enhancers with estrogen & mtorc genesets
Mixed_genesets_intersection = lapply(zz_msig, function(x) { 
  zz_estro = gain_enhancergene[gain_enhancergene$gene %in% msigdb_estro, ]
  zz_estro = gain_enhancergene[gain_enhancergene$enhancer %in% zz_estro$enhancer, ] # take ALL rows of enhancers with any estrogen genes
  zz_nonestro = gain_enhancergene[!gain_enhancergene$enhancer %in% zz_estro$enhancer, ] # get enhancers that don't have ANY estrogen genes
  zz_geneset = gain_enhancergene[gain_enhancergene$gene %in% x$GENE, ]
  zz_geneset = gain_enhancergene[gain_enhancergene$enhancer %in% zz_geneset$enhancer, ] # take ALL rows of enhancers with any geneset genes
  zz_nongeneset = gain_enhancergene[!gain_enhancergene$enhancer %in% zz_geneset$enhancer, ] # get enhancers that don't have ANY geneset genes
  
  zz_estro_geneset = zz_estro[zz_estro$gene %in% x$GENE, ]
  zz_estro_geneset = zz_estro[zz_estro$enhancer %in% zz_estro_geneset$enhancer, ] 
  zz_estro_nongeneset = zz_estro[!zz_estro$enhancer %in% zz_estro_geneset$enhancer, ]
  zz_nonestro_geneset = zz_nonestro[zz_nonestro$gene %in% x$GENE, ]
  zz_nonestro_geneset = zz_nonestro[zz_nonestro$enhancer %in% zz_nonestro_geneset$enhancer, ]
  zz_nonestro_nongeneset = zz_nonestro[!zz_nonestro$enhancer %in% zz_nonestro_geneset$enhancer, ]
  
  zz_fisher = fisher.test(matrix(c(length(unique(zz_estro_geneset$enhancer)), length(unique(zz_nonestro_geneset$enhancer)),
                                   length(unique(zz_estro_nongeneset$enhancer)), length(unique(zz_nonestro_nongeneset$enhancer))),nrow=2))

  return (data.frame(All_gain=length(unique(gain_enhancergene$enhancer)),
                     Estro=length(unique(zz_estro$enhancer)), NonEstro=length(unique(zz_nonestro$enhancer)),
                     Geneset=length(unique(zz_geneset$enhancer)), NonGeneset=length(unique(zz_nongeneset$enhancer)),
                     EstroGeneset=length(unique(zz_estro_geneset$enhancer)), EstroNonGeneset=length(unique(zz_estro_nongeneset$enhancer)),
                     NonEstroGeneset=length(unique(zz_nonestro_geneset$enhancer)), NonEstroNonGeneset=length(unique(zz_nonestro_nongeneset$enhancer)),
                     FisherP = zz_fisher$p.value))
})
Mixed_genesets_intersection = do.call(rbind, Mixed_genesets_intersection)
Mixed_genesets_intersection$FDR = p.adjust(Mixed_genesets_intersection$FisherP, method='BH')
Mixed_genesets_intersection = Mixed_genesets_intersection[order(Mixed_genesets_intersection$FDR),]
Mixed_genesets_intersection
  
```


```{r}
Fluke_genesets_intersection$Group="Fluke"
C4_genesets_intersection$Group="C4"
Mixed_genesets_intersection$Group="Mixed"
Fluke_genesets_intersection = Fluke_genesets_intersection[order(Fluke_genesets_intersection$FisherP),]
C4_genesets_intersection = C4_genesets_intersection[order(C4_genesets_intersection$FisherP),]
Mixed_genesets_intersection = Mixed_genesets_intersection[order(Mixed_genesets_intersection$FisherP),]

Fluke_genesets_intersection = Fluke_genesets_intersection[Fluke_genesets_intersection$FDR<.05,]
C4_genesets_intersection = C4_genesets_intersection[C4_genesets_intersection$FDR<.05,]
Mixed_genesets_intersection = Mixed_genesets_intersection[Mixed_genesets_intersection$FDR<.05,]

if (nrow(Fluke_genesets_intersection)>5) { Fluke_genesets_intersection = Fluke_genesets_intersection[1:5,]}
if (nrow(C4_genesets_intersection)>5) { C4_genesets_intersection = C4_genesets_intersection[1:5,]}
if (nrow(Mixed_genesets_intersection)>5) { Mixed_genesets_intersection = Mixed_genesets_intersection[1:5,]}


if (nrow(Fluke_genesets_intersection)>0) {
  Fluke_genesets_intersection$rowID = -1 : - nrow(Fluke_genesets_intersection)
}
if (nrow(C4_genesets_intersection)>0) {
  C4_genesets_intersection$rowID = -1 : - nrow(C4_genesets_intersection)
  C4_genesets_intersection$rowID = C4_genesets_intersection$rowID - nrow(Fluke_genesets_intersection) - 1
}
if (nrow(Mixed_genesets_intersection)>0) {
  Mixed_genesets_intersection$rowID = -1 : - nrow(Mixed_genesets_intersection)
  Mixed_genesets_intersection$rowID = Mixed_genesets_intersection$rowID - nrow(Fluke_genesets_intersection) - 1 - nrow(C4_genesets_intersection) - 1 
}


zz_plot = rbind(Fluke_genesets_intersection, C4_genesets_intersection, Mixed_genesets_intersection)
zz_plot$Description = rownames(zz_plot)
zz_plot$logFDR = - log10(zz_plot$FDR)
zz_plot

ggplot(zz_plot, aes(x=logFDR, y= rowID, size=EstroGeneset)) +
  geom_point(color="#e67050", alpha=.6)  +
  scale_size(range=c(5, 20)) +
  geom_vline(xintercept = -log10(.05), linetype="dashed", color="grey30") +
  geom_hline(yintercept = c(-nrow(Fluke_genesets_intersection)-1, -nrow(Fluke_genesets_intersection)-nrow(C4_genesets_intersection)-2)) + 
  ylab(NULL) + scale_y_continuous(breaks = seq(-1, -20, by = -1), minor_breaks = seq(-1, -20, by = -1), limits=c(-nrow(zz_plot)-2.5, -0.5))  +
  theme_bw(base_size=14) + xlab("- log10(FDR)") + scale_x_continuous(breaks = seq(5, 50, by = 5), minor_breaks = seq(5, 50, by = 5)) 
```



#### print number of gained estrogen enhancers per group
```{r}


###### Non Group-specific
### Fluke
gain_enhancergene = somatic_enhancers$Tumor_EGroupFluke_vs_Normal$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene$enhancergene = paste0(gain_enhancergene$enhancer, '|', gain_enhancergene$gene)
gain_enhancergene = gain_enhancergene[!duplicated(gain_enhancergene$enhancergene),]

# calculate overlap of gained enhancers with estrogen 
Fluke_estro = gain_enhancergene[gain_enhancergene$gene %in% msigdb_estro, ]
Fluke_estro = gain_enhancergene[gain_enhancergene$enhancer %in% Fluke_estro$enhancer, ] 
  
### C4
gain_enhancergene = somatic_enhancers$Tumor_EGroupC4_vs_Normal$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene$enhancergene = paste0(gain_enhancergene$enhancer, '|', gain_enhancergene$gene)
gain_enhancergene = gain_enhancergene[!duplicated(gain_enhancergene$enhancergene),]

# calculate overlap of gained enhancers with estrogen 
C4_estro = gain_enhancergene[gain_enhancergene$gene %in% msigdb_estro, ]
C4_estro = gain_enhancergene[gain_enhancergene$enhancer %in% C4_estro$enhancer, ] 
  
### Mixed
gain_enhancergene = somatic_enhancers$Tumor_EGroupMixed_vs_Normal$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene$enhancergene = paste0(gain_enhancergene$enhancer, '|', gain_enhancergene$gene)
gain_enhancergene = gain_enhancergene[!duplicated(gain_enhancergene$enhancergene),]

# calculate overlap of gained enhancers with estrogen 
Mixed_estro = gain_enhancergene[gain_enhancergene$gene %in% msigdb_estro, ]
Mixed_estro = gain_enhancergene[gain_enhancergene$enhancer %in% Mixed_estro$enhancer, ] 
  

print("Num gained estrogen enhancers, non group-specific")
print(paste0("Fluke: ", length(unique(Fluke_estro$enhancer))))
print(paste0("C4: ", length(unique(C4_estro$enhancer))))
print(paste0("Mixed: ", length(unique(Mixed_estro$enhancer))))




###### Group-specific
### Fluke
gain_enhancergene1 = somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupC4$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene1$enhancergene = paste0(gain_enhancergene1$enhancer, '|', gain_enhancergene1$gene)
gain_enhancergene1 = gain_enhancergene1[!duplicated(gain_enhancergene1$enhancergene),]
gain_enhancergene2 = somatic_enhancers_groupspec$Tumor_EGroupFluke_vs_Tumor_EGroupMixed$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene2$enhancergene = paste0(gain_enhancergene2$enhancer, '|', gain_enhancergene2$gene)
gain_enhancergene2 = gain_enhancergene2[!duplicated(gain_enhancergene2$enhancergene),]
gain_enhancergene = gain_enhancergene1[gain_enhancergene1$enhancergene %in% gain_enhancergene2$enhancergene,]

# calculate overlap of gained enhancers with estrogen 
Fluke_estro = gain_enhancergene[gain_enhancergene$gene %in% msigdb_estro, ]
Fluke_estro = gain_enhancergene[gain_enhancergene$enhancer %in% Fluke_estro$enhancer, ] 


### C4
gain_enhancergene1 = somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupFluke$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene1$enhancergene = paste0(gain_enhancergene1$enhancer, '|', gain_enhancergene1$gene)
gain_enhancergene1 = gain_enhancergene1[!duplicated(gain_enhancergene1$enhancergene),]
gain_enhancergene2 = somatic_enhancers_groupspec$Tumor_EGroupC4_vs_Tumor_EGroupMixed$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene2$enhancergene = paste0(gain_enhancergene2$enhancer, '|', gain_enhancergene2$gene)
gain_enhancergene2 = gain_enhancergene2[!duplicated(gain_enhancergene2$enhancergene),]
gain_enhancergene = gain_enhancergene1[gain_enhancergene1$enhancergene %in% gain_enhancergene2$enhancergene,]

# calculate overlap of gained enhancers with estrogen 
C4_estro = gain_enhancergene[gain_enhancergene$gene %in% msigdb_estro, ]
C4_estro = gain_enhancergene[gain_enhancergene$enhancer %in% C4_estro$enhancer, ] 


### Mixed
gain_enhancergene1 = somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupFluke$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene1$enhancergene = paste0(gain_enhancergene1$enhancer, '|', gain_enhancergene1$gene)
gain_enhancergene1 = gain_enhancergene1[!duplicated(gain_enhancergene1$enhancergene),]
gain_enhancergene2 = somatic_enhancers_groupspec$Tumor_EGroupMixed_vs_Tumor_EGroupC4$gain_exp_enhgain[, c("enhancer","gene")]
gain_enhancergene2$enhancergene = paste0(gain_enhancergene2$enhancer, '|', gain_enhancergene2$gene)
gain_enhancergene2 = gain_enhancergene2[!duplicated(gain_enhancergene2$enhancergene),]
gain_enhancergene = gain_enhancergene1[gain_enhancergene1$enhancergene %in% gain_enhancergene2$enhancergene,]

# calculate overlap of gained enhancers with estrogen 
Mixed_estro = gain_enhancergene[gain_enhancergene$gene %in% msigdb_estro, ]
Mixed_estro = gain_enhancergene[gain_enhancergene$enhancer %in% Mixed_estro$enhancer, ] 



print("Num gained estrogen enhancers, non group-specific")
print(paste0("Fluke: ", length(unique(Fluke_estro$enhancer))))
print(paste0("C4: ", length(unique(C4_estro$enhancer))))
print(paste0("Mixed: ", length(unique(Mixed_estro$enhancer))))


  
```






## Investigate estrogen response
```{r fig.width=12, fig.asp=1.5}

estro_resp_genes = unique(c(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_ESTROGEN_RESPONSE_EARLY","GENE"], msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_ESTROGEN_RESPONSE_LATE","GENE"])) # 299

tnfa_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_TNFA_SIGNALING_VIA_NFKB","GENE"], estro_resp_genes) # 18
il2_resp_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_IL2_STAT5_SIGNALING","GENE"], estro_resp_genes) # 15
inflam_resp_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_INFLAMMATORY_RESPONSE","GENE"], estro_resp_genes) # 5
tgfb_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_TGF_BETA_SIGNALING","GENE"], estro_resp_genes) # 5
il6_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_IL6_JAK_STAT3_SIGNALING","GENE"], estro_resp_genes) # 4
allo_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_ALLOGRAFT_REJECTION","GENE"], estro_resp_genes) # 3
ifng_resp_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_INTERFERON_GAMMA_RESPONSE","GENE"], estro_resp_genes) # 3
ifna_resp_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_INTERFERON_ALPHA_RESPONSE","GENE"], estro_resp_genes) # 2

mtorc1_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_MTORC1_SIGNALING","GENE"], estro_resp_genes) # 13
p53_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_P53_PATHWAY","GENE"], estro_resp_genes) # 12
pi3k_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_PI3K_AKT_MTOR_SIGNALING","GENE"], estro_resp_genes) # 3

myo_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_MYOGENESIS","GENE"], estro_resp_genes) # 13
emt_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION","GENE"], estro_resp_genes) # 10

glyco_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_GLYCOLYSIS","GENE"], estro_resp_genes) # 12
oxphos_genes = intersect(msigdb_H_term2gene[msigdb_H_term2gene$TERM=="HALLMARK_OXIDATIVE_PHOSPHORYLATION","GENE"], estro_resp_genes) # 3


### Clus1: get enhancer change vs normal. Set to 0: those with expr in wrong direction, or not in cell line
clus1_estro_enh = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts1, K27ac_dba2=K27ac_dba_deseq_contrasts1, contrasts=c(1,1), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts1, diffexprlist2=diffexpr_list_contrasts1, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=1, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=2, group_interest.cell=1, enh_present_thr=.5)
clus1_estro_enh = clus1_estro_enh[names(clus1_estro_enh) %in% estro_resp_genes] # 284, 78>2, 126>1, 130>0, 7 0's, 147<0, 17< -1, 1< -2

# msigdb_H_term2gene_zz = msigdb_H_term2gene[msigdb_H_term2gene$TERM != "ALL_GENES",]
# clus1_estro_enh_gsea = GSEA(geneList=clus1_estro_enh, minGSSize=1, maxGSSize=5000, TERM2GENE=msigdb_H_term2gene_zz, seed=12345, nPerm=1000, pvalueCutoff=1)

### Clus2: get enhancer change vs normal. Set to 0: those with expr in wrong direction, or not in cell line
clus2_estro_enh = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts2, K27ac_dba2=K27ac_dba_deseq_contrasts2, contrasts=c(1,1), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts2, diffexprlist2=diffexpr_list_contrasts2, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=1, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
clus2_estro_enh = clus2_estro_enh[names(clus2_estro_enh) %in% estro_resp_genes] # 284, 53>2, 141>1, 150>0, 10 0's, 124<0, 4< -1, 0< -2

### Clus3: get enhancer change vs normal. Set to 0: those with expr in wrong direction, or not in cell line
clus3_estro_enh = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts3, K27ac_dba2=K27ac_dba_deseq_contrasts3, contrasts=c(1,1), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts3, diffexprlist2=diffexpr_list_contrasts3, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=1, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
clus3_estro_enh = clus3_estro_enh[names(clus3_estro_enh) %in% estro_resp_genes] # 284, 38>2, 125>1, 128>0, 5 0's, 151<0, 16< -1, 1< -2

### Clus4: get enhancer change vs normal. Set to 0: those with expr in wrong direction, or not in cell line
clus4_estro_enh = get_gsea_genelist_multicomp_filterExpr_Cell(K27ac_dba1=K27ac_dba_deseq_contrasts4, K27ac_dba2=K27ac_dba_deseq_contrasts4, contrasts=c(1,1), group_interest=c(1,1),method=DBA_DESEQ2,                                             diffexprlist1=diffexpr_list_contrasts4, diffexprlist2=diffexpr_list_contrasts4, corr_data=target_map, filterEnh=NULL, filterExpr=T, 
                                            filterCell=1, K27ac_dba.cell=K27ac_cells_tisspeaks_dba, contrast.cell=3, group_interest.cell=1, enh_present_thr=.5)
clus4_estro_enh = clus4_estro_enh[names(clus4_estro_enh) %in% estro_resp_genes] # 284, 56>2, 110>1, 117>0, 5 0's, 162<0, 16< -1, 1< -2

###  cross with other genesets of interest
other_genesets = matrix(0, nrow=15, ncol=length(clus1_estro_enh))
colnames(other_genesets) = names(clus1_estro_enh)
rownames(other_genesets) = c("TNFA/NFKB", "IL2/STAT5", "INFLAM_RESP", "TGFB", "IL6/STAT3", "ALLOGRAFT", "IFNA", "IFNG","MYOGEN", "EMT", "MTORC1", "P53", "PI3K", "GLYCO", "OXPHOS")
other_genesets["TNFA/NFKB", colnames(other_genesets) %in% tnfa_genes] = 1
other_genesets["IL2/STAT5", colnames(other_genesets) %in% il2_resp_genes] = 1
other_genesets["INFLAM_RESP", colnames(other_genesets) %in% inflam_resp_genes] = 1
other_genesets["TGFB", colnames(other_genesets) %in% tgfb_genes] = 1
other_genesets["IL6/STAT3", colnames(other_genesets) %in% il6_genes] = 1
other_genesets["ALLOGRAFT", colnames(other_genesets) %in% allo_genes] = 1
other_genesets["IFNA", colnames(other_genesets) %in% ifng_resp_genes] = 1
other_genesets["IFNG", colnames(other_genesets) %in% ifna_resp_genes] = 1
other_genesets["MYOGEN", colnames(other_genesets) %in% myo_genes] = 1
other_genesets["EMT", colnames(other_genesets) %in% emt_genes] = 1
other_genesets["P53", colnames(other_genesets) %in% p53_genes] = 1
other_genesets["MTORC1", colnames(other_genesets) %in% mtorc1_genes] = 1
other_genesets["PI3K", colnames(other_genesets) %in% pi3k_genes] = 1
other_genesets["OXPHOS", colnames(other_genesets) %in% oxphos_genes] = 1
other_genesets["GLYCO", colnames(other_genesets) %in% glyco_genes] = 1
other_genesets = other_genesets[c(15:14, 13:11, 10:9, 8:1),]


# plot all, sorted individually
clus1_estro_enh = sort(clus1_estro_enh, decreasing = T) 
clus2_estro_enh = sort(clus2_estro_enh, decreasing = T) 
clus3_estro_enh = sort(clus3_estro_enh, decreasing = T) 
clus4_estro_enh = sort(clus4_estro_enh, decreasing = T)  
layout(mat=matrix(1:4, 4, 1, byrow = T))
barplot(clus1_estro_enh, border="blue", col="white", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus2_estro_enh, border="orange", col="white", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus3_estro_enh, border="green", col="white", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus4_estro_enh, border="red", col="white", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))

# sorted by clus1
clus1_estro_enh = sort(clus1_estro_enh, decreasing = T)
clus2_estro_enh = clus2_estro_enh[names(clus1_estro_enh)]
clus3_estro_enh = clus3_estro_enh[names(clus1_estro_enh)]
clus4_estro_enh = clus4_estro_enh[names(clus1_estro_enh)]
other_genesets = other_genesets[, names(clus1_estro_enh)]
layout(mat=matrix(1:5, 5, 1, byrow = T))
barplot(clus1_estro_enh, col=sapply(clus1_estro_enh-clus1_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="blue", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus2_estro_enh, col=sapply(clus2_estro_enh-clus1_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="orange", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus3_estro_enh, col=sapply(clus3_estro_enh-clus1_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="green", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus4_estro_enh, col=sapply(clus4_estro_enh-clus1_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="red", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
heatmap(other_genesets, Rowv=NA, Colv=NA, scale="none", col=c("azure2","black"))

# sorted by clus2
clus2_estro_enh = sort(clus2_estro_enh, decreasing = T)
clus1_estro_enh = clus1_estro_enh[names(clus2_estro_enh)]
clus3_estro_enh = clus3_estro_enh[names(clus2_estro_enh)]
clus4_estro_enh = clus4_estro_enh[names(clus2_estro_enh)]
other_genesets = other_genesets[, names(clus2_estro_enh)]
layout(mat=matrix(1:5, 5, 1, byrow = T))
barplot(clus1_estro_enh, col=sapply(clus1_estro_enh-clus2_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="blue", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus2_estro_enh, col=sapply(clus2_estro_enh-clus2_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="orange",ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus3_estro_enh, col=sapply(clus3_estro_enh-clus2_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="green", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus4_estro_enh, col=sapply(clus4_estro_enh-clus2_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="red", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
heatmap(other_genesets, Rowv=NA, Colv=NA, scale="none", col=c("azure2","black"))

# sorted by clus3
clus3_estro_enh = sort(clus3_estro_enh, decreasing = T)
clus1_estro_enh = clus1_estro_enh[names(clus3_estro_enh)]
clus2_estro_enh = clus2_estro_enh[names(clus3_estro_enh)]
clus4_estro_enh = clus4_estro_enh[names(clus3_estro_enh)]
other_genesets = other_genesets[, names(clus3_estro_enh)]
layout(mat=matrix(1:5, 5, 1, byrow = T))
barplot(clus1_estro_enh, col=sapply(clus1_estro_enh-clus3_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}),  border="blue", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus2_estro_enh, col=sapply(clus2_estro_enh-clus3_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="orange",ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus3_estro_enh, col=sapply(clus3_estro_enh-clus3_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="green",ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus4_estro_enh, col=sapply(clus4_estro_enh-clus3_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="red",ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
heatmap(other_genesets, Rowv=NA, Colv=NA, scale="none", col=c("azure2","black"))

# sorted by clus4
clus4_estro_enh = sort(clus4_estro_enh, decreasing = T)
clus1_estro_enh = clus1_estro_enh[names(clus4_estro_enh)]
clus2_estro_enh = clus2_estro_enh[names(clus4_estro_enh)]
clus3_estro_enh = clus3_estro_enh[names(clus4_estro_enh)]
other_genesets = other_genesets[, names(clus4_estro_enh)]
layout(mat=matrix(1:5, 5, 1, byrow = T))
barplot(clus1_estro_enh, col=sapply(clus1_estro_enh-clus4_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="blue", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus2_estro_enh, col=sapply(clus2_estro_enh-clus4_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="orange",ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus3_estro_enh, col=sapply(clus3_estro_enh-clus4_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="green", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
barplot(clus4_estro_enh, col=sapply(clus4_estro_enh-clus4_estro_enh, function(x){if(abs(x)>1){"black"} else{"white"}}), border="red", ylim=c(-5,5), names.arg="")
abline(h=c(1,-1))
heatmap(other_genesets, Rowv=NA, Colv=NA, scale="none", col=c("azure2","black"))
```






